<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.8.4" />
<title>pacmill.taxonomy.mothur_classify API documentation</title>
<meta name="description" content="Written by Lucas Sinclair.
MIT Licensed.
Contact at www.sinclair.bio" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.18.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.18.1/highlight.min.js" integrity="sha256-eOgo0OtLL4cdq7RdwRUiGKLX9XsIJ7nGhWEKbohmVAQ=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>pacmill.taxonomy.mothur_classify</code></h1>
</header>
<section id="section-intro">
<p>Written by Lucas Sinclair.
MIT Licensed.
Contact at www.sinclair.bio</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">#!/usr/bin/env python3
# -*- coding: utf-8 -*-

&#34;&#34;&#34;
Written by Lucas Sinclair.
MIT Licensed.
Contact at www.sinclair.bio
&#34;&#34;&#34;

# Built-in modules #
import os, multiprocessing, re

# First party modules #
from fasta import FASTA
from plumbing.check_cmd_found import check_cmd
from plumbing.cache           import property_cached
from autopaths.file_path      import FilePath
from autopaths.dir_path       import DirectoryPath

# Third party modules #
import sh

###############################################################################
class MothurClassify:
    &#34;&#34;&#34;
    Predicts taxonomic composition by using a reference database and calling
    `classify.seqs` in mothur as seen in:

    https://github.com/novigit/broCode/blob/master/pbamp/estimateTaxComposition.sh#L31

        mothur &#34;#classify.seqs(fasta=rnammer.$sample.prok.16S.both.fasta,
        reference=silva.nr_v128.align, taxonomy=silva.nr_v128.tax,
        relabund=T, processors=30)&#34;

    This is the method that is implemented by the RDP and is described by
    Wang et al.
    &#34;&#34;&#34;

    # Constants #
    short_name = &#39;mothur_classify&#39;
    long_name  = &#39;Mothur Version 1.42.1&#39;

    def __repr__(self):
        msg = &#39;&lt;%s object on &#34;%s&#34;&gt;&#39;
        return msg % (self.__class__.__name__, self.source.path)

    def __init__(self, source, database, dest_dir):
        # Source is the FASTA file containing OTU consensus sequences #
        self.source = FASTA(source)
        # Database is an object from the `seqsearch.databases` subpackage #
        self.database = database
        # Destination is a directory that contains all the results #
        self.dest_dir = DirectoryPath(dest_dir)

    #----------------------------- Installing --------------------------------#
    def check_installed(self):
        &#34;&#34;&#34;
        Try to determine if the database chosen is downloaded and
        accessible.
        &#34;&#34;&#34;
        if not self.database:
            msg  = &#34;The taxonomic database does not seem to be accessible.&#34;
            msg += &#34; More information follows.\n\n&#34;
            msg += self.database.__doc__
            raise Exception(msg)

    #-------------------------- Automatic paths ------------------------------#
    all_paths = &#34;&#34;&#34;
                /centers.fasta
                /assignments.txt
                /summary.txt
                /flipped.txt
                /stdout.txt
                /stderr.txt
                /taxa_tables/
                &#34;&#34;&#34;

    @property_cached
    def autopaths(self):
        &#34;&#34;&#34;
        The AutoPaths object is used for quickly assessing the filesystem paths
        of various file inputs/outputs and directories.
        See https://github.com/xapple/autopaths#autopaths-object
        &#34;&#34;&#34;
        from autopaths.auto_paths import AutoPaths
        return AutoPaths(self.dest_dir, self.all_paths)

    #------------------------------ Running ----------------------------------#
    def __call__(self, cpus=None, verbose=True):
        # Message #
        if verbose:
            message = &#34;Running `%s` taxonomy classification on &#39;%s&#39;&#34;
            print(message % (self.database.short_name, self.source))
        # Check the database is there #
        self.check_installed()
        # Check mothur is installed #
        check_cmd(&#39;mothur&#39;, True)
        # Remove the destination directory #
        self.dest_dir.remove()
        # Link our input OTU sequences to the destination directory #
        self.autopaths.centers.link_from(self.source)
        # Check version #
        assert &#34;1.42.1&#34; in sh.mothur(&#39;--version&#39;)
        # Make the long command as multiple strings #
        cmd = (&#34;#classify.seqs(&#34;, # The command
               &#34; fasta=%s,&#34;     , # The input file
               &#34; reference=%s,&#34; , # The database (was also called &#39;template&#39;)
               &#34; taxonomy=%s,&#34;  , # The taxonomy file
               &#34; processors=%s,&#34;, # The number of threads
               &#34; probs=F);&#34;)      # Disable the output of bootstrap values.
        # Turn it into a single string #
        cmd = &#39;&#39;.join(cmd)
        # Number of cores #
        if cpus is None: cpus = min(multiprocessing.cpu_count(), 32)
        # Format the string to contain all parameters #
        cmd = cmd % (self.autopaths.centers,
                     self.database.alignment,
                     self.database.taxonomy,
                     cpus)
        # Mothur pollutes with so much files, have to cwd #
        current_dir = os.getcwd()
        os.chdir(self.dest_dir)
        # Run the command on the input FASTA file #
        sh.mothur(cmd,
                  _out = self.autopaths.stdout.path,
                  _err = self.autopaths.stderr.path)
        # Restore current directory #
        os.chdir(current_dir)
        # Check output #
        if &#34;ERROR&#34; in self.autopaths.stdout.contents:
            msg  = &#34;The Mothur classify operation did not run correctly.&#34;
            msg += &#34;Please check the log files.&#34;
            raise Exception(msg)
        # Outputs #
        asgnmts = self.dest_dir + &#34;centers.%s.wang.taxonomy&#34;
        summary = self.dest_dir + &#34;centers.%s.wang.tax.summary&#34;
        flipped = self.dest_dir + &#34;centers.%s.wang.flip.accnos&#34;
        # Make into FilePath objects #
        asgnmts = FilePath(asgnmts % self.database.nickname)
        summary = FilePath(summary % self.database.nickname)
        flipped = FilePath(flipped % self.database.nickname)
        # Rename files #
        asgnmts.move_to(self.autopaths.assignments)
        summary.move_to(self.autopaths.summary)
        if flipped: flipped.move_to(self.autopaths.flipped)
        # Return #
        return self.dest_dir

    #------------------------------- Results ---------------------------------#
    def __bool__(self):
        &#34;&#34;&#34;
        Return True if the taxonomic software was run already and the results
        are stored on the filesystem. Return False if it was not yet run.
        &#34;&#34;&#34;
        return self.autopaths.assignments.exists

    @property_cached
    def results(self):
        # Check it was run #
        if not self:
            msg = &#34;You can&#39;t access results from the mothur &#39;%s&#39; taxonomic&#34; \
                  &#34; classification before running the tool.&#34;
            raise Exception(msg % self.database.short_name)
        # Return the results #
        return MothurClassifyResults(self.autopaths)

###############################################################################
class MothurClassifyResults:

    def __init__(self, autopaths):
        self.autopaths = autopaths

    @property_cached
    def assignments(self):
        &#34;&#34;&#34;
        Will parse the assignment file generated by mothur.
        Typically this attribute will return something like:

             &#39;OTU-3576&#39;: (&#39;Bacteria&#39;,
                          &#39;Bacteria_unclassified&#39;,
                          &#39;Bacteria_unclassified&#39;,
                          &#39;Bacteria_unclassified&#39;,
                          &#39;Bacteria_unclassified&#39;,
                          &#39;Bacteria_unclassified&#39;,
                          &#39;&#39;)
             &#39;OTU-3577&#39;: (&#39;Bacteria&#39;,
                          &#39;Firmicutes&#39;,
                          &#39;Clostridia&#39;,
                          &#39;Clostridiales&#39;,
                          &#39;Clostridiaceae_1&#39;,
                          &#39;Clostridiaceae_1_unclassified&#39;,
                          &#39;&#39;)
             &#39;OTU-4086&#39;: (&#39;unknown&#39;,
                          &#39;unknown_unclassified&#39;,
                          &#39;unknown_unclassified&#39;,
                          &#39;unknown_unclassified&#39;,
                          &#39;unknown_unclassified&#39;,
                          &#39;&#39;)
        &#34;&#34;&#34;
        # Initialize #
        result = {}
        with open(self.autopaths.assignments, &#39;r&#39;) as handle:
            for line in handle:
                # Get the OTU name and the assignment as a string #
                otu_name, species = line.split(&#39;\t&#39;)
                # Split the assignment into ranks #
                species = [i.strip(&#39;\n&#39;) for i in species.split(&#39;;&#39;)]
                # The OTU name matches what vsearch outputted in the FASTA #
                # And not what vsearch outputted in the TSV. We fix that #
                pattern  = &#39;\Acentroid=(.+);seqs=[0-9]+\Z&#39;
                otu_name = re.findall(pattern, otu_name)[0]
                # Assign that OTU to that tuple of ranks #
                result[otu_name] = tuple(species)
        # Return #
        return result

    @property_cached
    def count_unassigned(self):
        &#34;&#34;&#34;Will count how many did not get a prediction at each level.&#34;&#34;&#34;
        # Load the assignment values created by mothur #
        vals = list(self.assignments.values())
        # Calculate for each position in the tree of life #
        return [
            sum((1 for x in vals if x[0] == &#39;unknown&#39;)),      # Domain
            sum((1 for x in vals if &#39;unclassified&#39; in x[1])), # Phylum
            sum((1 for x in vals if &#39;unclassified&#39; in x[2])), # Class
            sum((1 for x in vals if &#39;unclassified&#39; in x[3])), # Order
            sum((1 for x in vals if &#39;unclassified&#39; in x[4])), # Family
            sum((1 for x in vals if &#39;unclassified&#39; in x[5])), # Genus
            sum((1 for x in vals if x[6] == &#39;&#39;)),             # Species
        ]

    @property_cached
    def count_assigned(self):
        &#34;&#34;&#34;Will count how many did get a prediction at each level.&#34;&#34;&#34;
        return [len(self.assignments) - x for x in self.count_unassigned]</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="pacmill.taxonomy.mothur_classify.MothurClassify"><code class="flex name class">
<span>class <span class="ident">MothurClassify</span></span>
<span>(</span><span>source, database, dest_dir)</span>
</code></dt>
<dd>
<div class="desc"><p>Predicts taxonomic composition by using a reference database and calling
<code>classify.seqs</code> in mothur as seen in:</p>
<p><a href="https://github.com/novigit/broCode/blob/master/pbamp/estimateTaxComposition.sh#L31">https://github.com/novigit/broCode/blob/master/pbamp/estimateTaxComposition.sh#L31</a></p>
<pre><code>mothur "#classify.seqs(fasta=rnammer.$sample.prok.16S.both.fasta,
reference=silva.nr_v128.align, taxonomy=silva.nr_v128.tax,
relabund=T, processors=30)"
</code></pre>
<p>This is the method that is implemented by the RDP and is described by
Wang et al.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class MothurClassify:
    &#34;&#34;&#34;
    Predicts taxonomic composition by using a reference database and calling
    `classify.seqs` in mothur as seen in:

    https://github.com/novigit/broCode/blob/master/pbamp/estimateTaxComposition.sh#L31

        mothur &#34;#classify.seqs(fasta=rnammer.$sample.prok.16S.both.fasta,
        reference=silva.nr_v128.align, taxonomy=silva.nr_v128.tax,
        relabund=T, processors=30)&#34;

    This is the method that is implemented by the RDP and is described by
    Wang et al.
    &#34;&#34;&#34;

    # Constants #
    short_name = &#39;mothur_classify&#39;
    long_name  = &#39;Mothur Version 1.42.1&#39;

    def __repr__(self):
        msg = &#39;&lt;%s object on &#34;%s&#34;&gt;&#39;
        return msg % (self.__class__.__name__, self.source.path)

    def __init__(self, source, database, dest_dir):
        # Source is the FASTA file containing OTU consensus sequences #
        self.source = FASTA(source)
        # Database is an object from the `seqsearch.databases` subpackage #
        self.database = database
        # Destination is a directory that contains all the results #
        self.dest_dir = DirectoryPath(dest_dir)

    #----------------------------- Installing --------------------------------#
    def check_installed(self):
        &#34;&#34;&#34;
        Try to determine if the database chosen is downloaded and
        accessible.
        &#34;&#34;&#34;
        if not self.database:
            msg  = &#34;The taxonomic database does not seem to be accessible.&#34;
            msg += &#34; More information follows.\n\n&#34;
            msg += self.database.__doc__
            raise Exception(msg)

    #-------------------------- Automatic paths ------------------------------#
    all_paths = &#34;&#34;&#34;
                /centers.fasta
                /assignments.txt
                /summary.txt
                /flipped.txt
                /stdout.txt
                /stderr.txt
                /taxa_tables/
                &#34;&#34;&#34;

    @property_cached
    def autopaths(self):
        &#34;&#34;&#34;
        The AutoPaths object is used for quickly assessing the filesystem paths
        of various file inputs/outputs and directories.
        See https://github.com/xapple/autopaths#autopaths-object
        &#34;&#34;&#34;
        from autopaths.auto_paths import AutoPaths
        return AutoPaths(self.dest_dir, self.all_paths)

    #------------------------------ Running ----------------------------------#
    def __call__(self, cpus=None, verbose=True):
        # Message #
        if verbose:
            message = &#34;Running `%s` taxonomy classification on &#39;%s&#39;&#34;
            print(message % (self.database.short_name, self.source))
        # Check the database is there #
        self.check_installed()
        # Check mothur is installed #
        check_cmd(&#39;mothur&#39;, True)
        # Remove the destination directory #
        self.dest_dir.remove()
        # Link our input OTU sequences to the destination directory #
        self.autopaths.centers.link_from(self.source)
        # Check version #
        assert &#34;1.42.1&#34; in sh.mothur(&#39;--version&#39;)
        # Make the long command as multiple strings #
        cmd = (&#34;#classify.seqs(&#34;, # The command
               &#34; fasta=%s,&#34;     , # The input file
               &#34; reference=%s,&#34; , # The database (was also called &#39;template&#39;)
               &#34; taxonomy=%s,&#34;  , # The taxonomy file
               &#34; processors=%s,&#34;, # The number of threads
               &#34; probs=F);&#34;)      # Disable the output of bootstrap values.
        # Turn it into a single string #
        cmd = &#39;&#39;.join(cmd)
        # Number of cores #
        if cpus is None: cpus = min(multiprocessing.cpu_count(), 32)
        # Format the string to contain all parameters #
        cmd = cmd % (self.autopaths.centers,
                     self.database.alignment,
                     self.database.taxonomy,
                     cpus)
        # Mothur pollutes with so much files, have to cwd #
        current_dir = os.getcwd()
        os.chdir(self.dest_dir)
        # Run the command on the input FASTA file #
        sh.mothur(cmd,
                  _out = self.autopaths.stdout.path,
                  _err = self.autopaths.stderr.path)
        # Restore current directory #
        os.chdir(current_dir)
        # Check output #
        if &#34;ERROR&#34; in self.autopaths.stdout.contents:
            msg  = &#34;The Mothur classify operation did not run correctly.&#34;
            msg += &#34;Please check the log files.&#34;
            raise Exception(msg)
        # Outputs #
        asgnmts = self.dest_dir + &#34;centers.%s.wang.taxonomy&#34;
        summary = self.dest_dir + &#34;centers.%s.wang.tax.summary&#34;
        flipped = self.dest_dir + &#34;centers.%s.wang.flip.accnos&#34;
        # Make into FilePath objects #
        asgnmts = FilePath(asgnmts % self.database.nickname)
        summary = FilePath(summary % self.database.nickname)
        flipped = FilePath(flipped % self.database.nickname)
        # Rename files #
        asgnmts.move_to(self.autopaths.assignments)
        summary.move_to(self.autopaths.summary)
        if flipped: flipped.move_to(self.autopaths.flipped)
        # Return #
        return self.dest_dir

    #------------------------------- Results ---------------------------------#
    def __bool__(self):
        &#34;&#34;&#34;
        Return True if the taxonomic software was run already and the results
        are stored on the filesystem. Return False if it was not yet run.
        &#34;&#34;&#34;
        return self.autopaths.assignments.exists

    @property_cached
    def results(self):
        # Check it was run #
        if not self:
            msg = &#34;You can&#39;t access results from the mothur &#39;%s&#39; taxonomic&#34; \
                  &#34; classification before running the tool.&#34;
            raise Exception(msg % self.database.short_name)
        # Return the results #
        return MothurClassifyResults(self.autopaths)</code></pre>
</details>
<h3>Class variables</h3>
<dl>
<dt id="pacmill.taxonomy.mothur_classify.MothurClassify.all_paths"><code class="name">var <span class="ident">all_paths</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="pacmill.taxonomy.mothur_classify.MothurClassify.long_name"><code class="name">var <span class="ident">long_name</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="pacmill.taxonomy.mothur_classify.MothurClassify.short_name"><code class="name">var <span class="ident">short_name</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
</dl>
<h3>Instance variables</h3>
<dl>
<dt id="pacmill.taxonomy.mothur_classify.MothurClassify.autopaths"><code class="name">var <span class="ident">autopaths</span></code></dt>
<dd>
<div class="desc"><p>The AutoPaths object is used for quickly assessing the filesystem paths
of various file inputs/outputs and directories.
See <a href="https://github.com/xapple/autopaths#autopaths-object">https://github.com/xapple/autopaths#autopaths-object</a></p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def __get__(self, instance, owner):
    &#34;&#34;&#34;
    If you see the current source code in a seemingly unrelated part of
    an auto-generated documentation, it means the program making the
    documentation was unable to correctly traverse a decorated property.
    &#34;&#34;&#34;
    # For debugging purposes #
    if False: print(&#34;-&gt; property cached `%s`&#34; % instance)
    # If called from a class #
    if instance is None: return self
    # Does a cache exist for this instance? #
    self.check_cache(instance)
    # Is the answer in the cache? #
    if self.name in instance.__cache__: return instance.__cache__[self.name]
    # If not we will compute it #
    if inspect.isgeneratorfunction(self.func): result = tuple(self.func(instance))
    else:                                      result = self.func(instance)
    # Let&#39;s store the answer for later #
    instance.__cache__[self.name] = result
    # Return #
    return result</code></pre>
</details>
</dd>
<dt id="pacmill.taxonomy.mothur_classify.MothurClassify.results"><code class="name">var <span class="ident">results</span></code></dt>
<dd>
<div class="desc"><p>If you see the current source code in a seemingly unrelated part of
an auto-generated documentation, it means the program making the
documentation was unable to correctly traverse a decorated property.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def __get__(self, instance, owner):
    &#34;&#34;&#34;
    If you see the current source code in a seemingly unrelated part of
    an auto-generated documentation, it means the program making the
    documentation was unable to correctly traverse a decorated property.
    &#34;&#34;&#34;
    # For debugging purposes #
    if False: print(&#34;-&gt; property cached `%s`&#34; % instance)
    # If called from a class #
    if instance is None: return self
    # Does a cache exist for this instance? #
    self.check_cache(instance)
    # Is the answer in the cache? #
    if self.name in instance.__cache__: return instance.__cache__[self.name]
    # If not we will compute it #
    if inspect.isgeneratorfunction(self.func): result = tuple(self.func(instance))
    else:                                      result = self.func(instance)
    # Let&#39;s store the answer for later #
    instance.__cache__[self.name] = result
    # Return #
    return result</code></pre>
</details>
</dd>
</dl>
<h3>Methods</h3>
<dl>
<dt id="pacmill.taxonomy.mothur_classify.MothurClassify.check_installed"><code class="name flex">
<span>def <span class="ident">check_installed</span></span>(<span>self)</span>
</code></dt>
<dd>
<div class="desc"><p>Try to determine if the database chosen is downloaded and
accessible.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def check_installed(self):
    &#34;&#34;&#34;
    Try to determine if the database chosen is downloaded and
    accessible.
    &#34;&#34;&#34;
    if not self.database:
        msg  = &#34;The taxonomic database does not seem to be accessible.&#34;
        msg += &#34; More information follows.\n\n&#34;
        msg += self.database.__doc__
        raise Exception(msg)</code></pre>
</details>
</dd>
</dl>
</dd>
<dt id="pacmill.taxonomy.mothur_classify.MothurClassifyResults"><code class="flex name class">
<span>class <span class="ident">MothurClassifyResults</span></span>
<span>(</span><span>autopaths)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class MothurClassifyResults:

    def __init__(self, autopaths):
        self.autopaths = autopaths

    @property_cached
    def assignments(self):
        &#34;&#34;&#34;
        Will parse the assignment file generated by mothur.
        Typically this attribute will return something like:

             &#39;OTU-3576&#39;: (&#39;Bacteria&#39;,
                          &#39;Bacteria_unclassified&#39;,
                          &#39;Bacteria_unclassified&#39;,
                          &#39;Bacteria_unclassified&#39;,
                          &#39;Bacteria_unclassified&#39;,
                          &#39;Bacteria_unclassified&#39;,
                          &#39;&#39;)
             &#39;OTU-3577&#39;: (&#39;Bacteria&#39;,
                          &#39;Firmicutes&#39;,
                          &#39;Clostridia&#39;,
                          &#39;Clostridiales&#39;,
                          &#39;Clostridiaceae_1&#39;,
                          &#39;Clostridiaceae_1_unclassified&#39;,
                          &#39;&#39;)
             &#39;OTU-4086&#39;: (&#39;unknown&#39;,
                          &#39;unknown_unclassified&#39;,
                          &#39;unknown_unclassified&#39;,
                          &#39;unknown_unclassified&#39;,
                          &#39;unknown_unclassified&#39;,
                          &#39;&#39;)
        &#34;&#34;&#34;
        # Initialize #
        result = {}
        with open(self.autopaths.assignments, &#39;r&#39;) as handle:
            for line in handle:
                # Get the OTU name and the assignment as a string #
                otu_name, species = line.split(&#39;\t&#39;)
                # Split the assignment into ranks #
                species = [i.strip(&#39;\n&#39;) for i in species.split(&#39;;&#39;)]
                # The OTU name matches what vsearch outputted in the FASTA #
                # And not what vsearch outputted in the TSV. We fix that #
                pattern  = &#39;\Acentroid=(.+);seqs=[0-9]+\Z&#39;
                otu_name = re.findall(pattern, otu_name)[0]
                # Assign that OTU to that tuple of ranks #
                result[otu_name] = tuple(species)
        # Return #
        return result

    @property_cached
    def count_unassigned(self):
        &#34;&#34;&#34;Will count how many did not get a prediction at each level.&#34;&#34;&#34;
        # Load the assignment values created by mothur #
        vals = list(self.assignments.values())
        # Calculate for each position in the tree of life #
        return [
            sum((1 for x in vals if x[0] == &#39;unknown&#39;)),      # Domain
            sum((1 for x in vals if &#39;unclassified&#39; in x[1])), # Phylum
            sum((1 for x in vals if &#39;unclassified&#39; in x[2])), # Class
            sum((1 for x in vals if &#39;unclassified&#39; in x[3])), # Order
            sum((1 for x in vals if &#39;unclassified&#39; in x[4])), # Family
            sum((1 for x in vals if &#39;unclassified&#39; in x[5])), # Genus
            sum((1 for x in vals if x[6] == &#39;&#39;)),             # Species
        ]

    @property_cached
    def count_assigned(self):
        &#34;&#34;&#34;Will count how many did get a prediction at each level.&#34;&#34;&#34;
        return [len(self.assignments) - x for x in self.count_unassigned]</code></pre>
</details>
<h3>Instance variables</h3>
<dl>
<dt id="pacmill.taxonomy.mothur_classify.MothurClassifyResults.assignments"><code class="name">var <span class="ident">assignments</span></code></dt>
<dd>
<div class="desc"><p>Will parse the assignment file generated by mothur.
Typically this attribute will return something like:</p>
<pre><code> 'OTU-3576': ('Bacteria',
              'Bacteria_unclassified',
              'Bacteria_unclassified',
              'Bacteria_unclassified',
              'Bacteria_unclassified',
              'Bacteria_unclassified',
              '')
 'OTU-3577': ('Bacteria',
              'Firmicutes',
              'Clostridia',
              'Clostridiales',
              'Clostridiaceae_1',
              'Clostridiaceae_1_unclassified',
              '')
 'OTU-4086': ('unknown',
              'unknown_unclassified',
              'unknown_unclassified',
              'unknown_unclassified',
              'unknown_unclassified',
              '')
</code></pre></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def __get__(self, instance, owner):
    &#34;&#34;&#34;
    If you see the current source code in a seemingly unrelated part of
    an auto-generated documentation, it means the program making the
    documentation was unable to correctly traverse a decorated property.
    &#34;&#34;&#34;
    # For debugging purposes #
    if False: print(&#34;-&gt; property cached `%s`&#34; % instance)
    # If called from a class #
    if instance is None: return self
    # Does a cache exist for this instance? #
    self.check_cache(instance)
    # Is the answer in the cache? #
    if self.name in instance.__cache__: return instance.__cache__[self.name]
    # If not we will compute it #
    if inspect.isgeneratorfunction(self.func): result = tuple(self.func(instance))
    else:                                      result = self.func(instance)
    # Let&#39;s store the answer for later #
    instance.__cache__[self.name] = result
    # Return #
    return result</code></pre>
</details>
</dd>
<dt id="pacmill.taxonomy.mothur_classify.MothurClassifyResults.count_assigned"><code class="name">var <span class="ident">count_assigned</span></code></dt>
<dd>
<div class="desc"><p>Will count how many did get a prediction at each level.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def __get__(self, instance, owner):
    &#34;&#34;&#34;
    If you see the current source code in a seemingly unrelated part of
    an auto-generated documentation, it means the program making the
    documentation was unable to correctly traverse a decorated property.
    &#34;&#34;&#34;
    # For debugging purposes #
    if False: print(&#34;-&gt; property cached `%s`&#34; % instance)
    # If called from a class #
    if instance is None: return self
    # Does a cache exist for this instance? #
    self.check_cache(instance)
    # Is the answer in the cache? #
    if self.name in instance.__cache__: return instance.__cache__[self.name]
    # If not we will compute it #
    if inspect.isgeneratorfunction(self.func): result = tuple(self.func(instance))
    else:                                      result = self.func(instance)
    # Let&#39;s store the answer for later #
    instance.__cache__[self.name] = result
    # Return #
    return result</code></pre>
</details>
</dd>
<dt id="pacmill.taxonomy.mothur_classify.MothurClassifyResults.count_unassigned"><code class="name">var <span class="ident">count_unassigned</span></code></dt>
<dd>
<div class="desc"><p>Will count how many did not get a prediction at each level.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def __get__(self, instance, owner):
    &#34;&#34;&#34;
    If you see the current source code in a seemingly unrelated part of
    an auto-generated documentation, it means the program making the
    documentation was unable to correctly traverse a decorated property.
    &#34;&#34;&#34;
    # For debugging purposes #
    if False: print(&#34;-&gt; property cached `%s`&#34; % instance)
    # If called from a class #
    if instance is None: return self
    # Does a cache exist for this instance? #
    self.check_cache(instance)
    # Is the answer in the cache? #
    if self.name in instance.__cache__: return instance.__cache__[self.name]
    # If not we will compute it #
    if inspect.isgeneratorfunction(self.func): result = tuple(self.func(instance))
    else:                                      result = self.func(instance)
    # Let&#39;s store the answer for later #
    instance.__cache__[self.name] = result
    # Return #
    return result</code></pre>
</details>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="pacmill.taxonomy" href="index.html">pacmill.taxonomy</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="pacmill.taxonomy.mothur_classify.MothurClassify" href="#pacmill.taxonomy.mothur_classify.MothurClassify">MothurClassify</a></code></h4>
<ul class="two-column">
<li><code><a title="pacmill.taxonomy.mothur_classify.MothurClassify.all_paths" href="#pacmill.taxonomy.mothur_classify.MothurClassify.all_paths">all_paths</a></code></li>
<li><code><a title="pacmill.taxonomy.mothur_classify.MothurClassify.autopaths" href="#pacmill.taxonomy.mothur_classify.MothurClassify.autopaths">autopaths</a></code></li>
<li><code><a title="pacmill.taxonomy.mothur_classify.MothurClassify.check_installed" href="#pacmill.taxonomy.mothur_classify.MothurClassify.check_installed">check_installed</a></code></li>
<li><code><a title="pacmill.taxonomy.mothur_classify.MothurClassify.long_name" href="#pacmill.taxonomy.mothur_classify.MothurClassify.long_name">long_name</a></code></li>
<li><code><a title="pacmill.taxonomy.mothur_classify.MothurClassify.results" href="#pacmill.taxonomy.mothur_classify.MothurClassify.results">results</a></code></li>
<li><code><a title="pacmill.taxonomy.mothur_classify.MothurClassify.short_name" href="#pacmill.taxonomy.mothur_classify.MothurClassify.short_name">short_name</a></code></li>
</ul>
</li>
<li>
<h4><code><a title="pacmill.taxonomy.mothur_classify.MothurClassifyResults" href="#pacmill.taxonomy.mothur_classify.MothurClassifyResults">MothurClassifyResults</a></code></h4>
<ul class="">
<li><code><a title="pacmill.taxonomy.mothur_classify.MothurClassifyResults.assignments" href="#pacmill.taxonomy.mothur_classify.MothurClassifyResults.assignments">assignments</a></code></li>
<li><code><a title="pacmill.taxonomy.mothur_classify.MothurClassifyResults.count_assigned" href="#pacmill.taxonomy.mothur_classify.MothurClassifyResults.count_assigned">count_assigned</a></code></li>
<li><code><a title="pacmill.taxonomy.mothur_classify.MothurClassifyResults.count_unassigned" href="#pacmill.taxonomy.mothur_classify.MothurClassifyResults.count_unassigned">count_unassigned</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.8.4</a>.</p>
</footer>
</body>
</html>