<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.9.2" />
<title>pacmill.taxonomy.ncbi_blast API documentation</title>
<meta name="description" content="Written by Lucas Sinclair.
MIT Licensed.
Contact at www.sinclair.bio" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>pacmill.taxonomy.ncbi_blast</code></h1>
</header>
<section id="section-intro">
<p>Written by Lucas Sinclair.
MIT Licensed.
Contact at www.sinclair.bio</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">#!/usr/bin/env python3
# -*- coding: utf-8 -*-

&#34;&#34;&#34;
Written by Lucas Sinclair.
MIT Licensed.
Contact at www.sinclair.bio
&#34;&#34;&#34;

# Built-in modules #
import re

# First party modules #
from fasta import FASTA
from autopaths.dir_path       import DirectoryPath
from plumbing.cache           import property_cached
from plumbing.check_cmd_found import check_cmd

# Third party modules #
import pandas

###############################################################################
class BlastClassify:
    &#34;&#34;&#34;
    Use BLAST and the NCBI 16S database to identify close matches between
    OTU centroid sequences and cultured bacterial/archaeal 16 sequences.
    &#34;&#34;&#34;

    # Constants #
    short_name = &#39;ncbi_blast&#39;
    long_name  = &#39;BLAST against the NCBI 16S database&#39;

    # The database #
    db_short_name   = &#34;ncbi_16s_rna&#34;
    db_long_name    = &#39;The NCBI 16S RNA database&#39;

    # Parameters #
    min_e_value        = 1e-5
    min_perc_identity  = 97
    max_target_seqs    = 5

    def __repr__(self):
        msg = &#39;&lt;%s object on &#34;%s&#34;&gt;&#39;
        return msg % (self.__class__.__name__, self.source.path)

    def __init__(self, source, database, dest_dir, proj):
        # Source is the FASTA file containing OTU consensus sequences #
        self.source = FASTA(source)
        # Database is an object from the `seqsearch.databases` subpackage #
        self.database = database
        # Destination is a directory that contains all the results #
        self.dest_dir = DirectoryPath(dest_dir)
        # Keep a reference to the project #
        self.proj = proj

    #----------------------------- Installing --------------------------------#
    apt_packages = [&#39;ncbi-blast+&#39;]

    def check_installed(self, exception=True):
        &#34;&#34;&#34;
        Try to determine if the BLAST software is installed and accessible.
        Also try to determine if the database chosen is downloaded and
        accessible.
        &#34;&#34;&#34;
        # The database #
        if not self.database:
            msg  = &#34;The taxonomic database does not seem to be accessible.&#34;
            msg += &#34; More information follows.\n\n&#34;
            msg += self.database.__doc__
            raise Exception(msg)
        # The command #
        return check_cmd(&#39;blastn&#39;, exception)

    #-------------------------- Automatic paths ------------------------------#
    all_paths = &#34;&#34;&#34;
                /blast/hits.xml
                /blast/stdout.txt
                /blast/stderr.txt
                /summary/single_hit.xlsx
                /summary/five_hits.xlsx
                /report.pdf
                &#34;&#34;&#34;

    @property_cached
    def autopaths(self):
        &#34;&#34;&#34;
        The AutoPaths object is used for quickly assessing the filesystem paths
        of various file inputs/outputs and directories.
        See https://github.com/xapple/autopaths#autopaths-object
        &#34;&#34;&#34;
        from autopaths.auto_paths import AutoPaths
        return AutoPaths(self.dest_dir, self.all_paths)

    #------------------------------ Running ----------------------------------#
    @property_cached
    def seq_search(self):
        &#34;&#34;&#34;
        The SeqSearch object is used for quickly launching BLAST search queries
        against a database of choice starting from a FASTA object.
        See https://github.com/xapple/seqsearch
        &#34;&#34;&#34;
        # Import #
        from seqsearch.search import SeqSearch
        # Extra options for the BLAST search #
        params = {&#39;-evalue&#39;:          self.min_e_value,
                  &#39;-perc_identity&#39;:   self.min_perc_identity,
                  &#39;-max_target_seqs&#39;: self.max_target_seqs}
        # Create the object #
        return SeqSearch(input_fasta = self.source,
                         database    = self.database,
                         seq_type    = &#39;nucl&#39;,
                         out_path    = self.autopaths.hits,
                         _out        = self.autopaths.stdout,
                         _err        = self.autopaths.stderr,
                         params      = params)

    def __call__(self, cpus=None, verbose=True):
        # Message #
        if verbose:
            message = &#34;Running `%s` taxonomy classification on &#39;%s&#39;.&#34;
            print(message % (self.short_name, self.source))
        # Check blast is installed #
        self.check_installed()
        # Run #
        self.seq_search.run()
        # Message #
        if verbose:
            message = &#34;Generating BLAST summary Excel files at &#39;%s&#39;.&#34;
            print(message % self.autopaths.summary_dir)
        # Generate summary lists #
        one  = self.all_otus_df(1)
        five = self.all_otus_df(5)
        # Create excel reports #
        one.to_excel(self.autopaths.xlsx_single_hit.path, index=False)
        five.to_excel(self.autopaths.xlsx_five_hits.path, index=False)

    #------------------------------- Results ---------------------------------#
    def __bool__(self):
        &#34;&#34;&#34;
        Return True if the taxonomic software was run already and the results
        are stored on the filesystem. Return False if it was not yet run.
        &#34;&#34;&#34;
        return self.autopaths.hits.exists

    @property_cached
    def results(self):
        # Check it was run #
        if not self:
            msg = &#34;You can&#39;t access results from the BLAST taxonomic&#34; \
                  &#34; classification before running the tool.&#34;
            raise Exception(msg)
        # Return the results #
        return BlastResults(self)

    #------------------------------ Dataframes -------------------------------#
    # Define columns needed #
    columns = [&#39;abundance&#39;, &#39;otu_id&#39;, &#39;description&#39;, &#39;score&#39;, &#39;e_value&#39;,
               &#39;accession&#39;, &#39;length&#39;, &#39;cover&#39;, &#39;identity&#39;]

    def all_otus_df(self, num_hits_per_otu=5):
        &#34;&#34;&#34;
        Returns a dataframe object summarizing all the top hits for all OTUs.
        The results looks like:

        abundance       otu_id  hit_number  ...  accession  length  identity
             4233    small_1:0           1  ...  NR_113957    1475     99.9%
             4233    small_1:0           2  ...  NR_036904    1472    100.0%
             4233    small_1:0           3  ...  NR_113405    1488     99.5%
             4233    small_1:0           4  ...  NR_118997    1524     98.6%
        &#34;&#34;&#34;
        # Parse the BLAST XML output file with BioPython #
        all_otus_hits = self.seq_search.results
        # Produce one mini-dataframe for every OTU #
        all_otus_dfs = [self.one_otu_df(record, num_hits_per_otu)
                        for record in all_otus_hits]
        # Case there are no hits in any OTUs at all #
        if all(df is None for df in all_otus_dfs):
            return pandas.DataFrame()
        # Combine them all into a big dataframe #
        df = pandas.concat(all_otus_dfs)
        # Keep the hit number as a seperate column #
        df = df.reset_index()
        col = df.pop(&#34;index&#34;)
        df.insert(2, &#39;hit_number&#39;, col)
        df[&#39;hit_number&#39;] += 1
        # Sort by abundance but keep OTUs grouped of course #
        df = df.sort_values(by=[&#39;abundance&#39;, &#39;otu_id&#39;, &#39;hit_number&#39;],
                            ascending=[False, False, True])
        # Return #
        return df

    def one_otu_df(self, record, max_hits=5):
        &#34;&#34;&#34;
        Returns a dataframe object summarizing the top hits for a given OTU.

        The record object is of type: &lt;Bio.Blast.Record.Blast&gt;

        The title looks like: &#34;gi|631252759|ref|NR_113957.1| Staphylococcus...&#34;
        The query looks like: &#34;centroid=small_1:0;seqs=4233&#34;

        To calculate the percentage identity see:

            https://github.com/peterjc/galaxy_blast/blob/master/tools/
            ncbi_blast_plus/blastxml_to_tabular.py#L230
        &#34;&#34;&#34;
        # Case there are no hits #
        if not record.descriptions: return
        # Get OTU name and its abundance #
        pattern = &#39;\Acentroid=(.+);seqs=([0-9]+)\Z&#39;
        otu, abund = re.findall(pattern, record.query)[0]
        # Cast to integer #
        abund = int(abund)
        # The description field contains lots of info we don&#39;t need #
        shorten = lambda s: s.split(&#39;|&#39;)[-1].strip()
        # Generate one line per hit #
        def one_row_per_hit():
            for align, desc in zip(record.alignments, record.descriptions):
                # Calculate all parameters #
                score       = desc.score
                e_value     = desc.e
                acc         = desc.accession
                title       = shorten(desc.title)
                hsps        = align.hsps[0]
                ident       = hsps.identities
                # The lengths available #
                qry_length  = record.query_length
                hit_length  = align.length
                algn_length = hsps.align_length
                # The ratios defined #
                p_cover = float(qry_length) / float(algn_length)
                p_ident = float(ident)      / float(algn_length)
                # Format percentages as strings #
                p_cover = &#34;%0.1f%%&#34; % (100 * p_cover)
                p_ident = &#34;%0.1f%%&#34; % (100 * p_ident)
                # Yield one row #
                yield abund, otu, title, score, e_value, acc,\
                      algn_length, p_cover, p_ident
        # Convert to dataframe #
        df = pandas.DataFrame(one_row_per_hit(), columns=self.columns)
        # Sort by E-value and also by score #
        df = df.sort_values(by=[&#39;e_value&#39;, &#39;score&#39;], ascending=False)
        # Filter and take only the top hits #
        df = df.head(max_hits)
        # Return #
        return df

###############################################################################
class BlastResults:

    def __init__(self, tax):
        self.tax = tax

    @property_cached
    def report(self):
        &#34;&#34;&#34;
        The NcbiBlastReport object is used to produce a PDF document containing
        all the information that concern this particular taxonomic
        classification strategy.
        &#34;&#34;&#34;
        # Imports #
        from pacmill.reports.ncbi_blast import NcbiBlastReport
        # Output directory #
        path = self.tax.proj.autopaths.taxonomy_dir + &#39;reports/&#39;
        path = path + self.tax.short_name + &#39;.pdf&#39;
        # Return #
        return NcbiBlastReport(self.tax, path)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="pacmill.taxonomy.ncbi_blast.BlastClassify"><code class="flex name class">
<span>class <span class="ident">BlastClassify</span></span>
<span>(</span><span>source, database, dest_dir, proj)</span>
</code></dt>
<dd>
<div class="desc"><p>Use BLAST and the NCBI 16S database to identify close matches between
OTU centroid sequences and cultured bacterial/archaeal 16 sequences.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class BlastClassify:
    &#34;&#34;&#34;
    Use BLAST and the NCBI 16S database to identify close matches between
    OTU centroid sequences and cultured bacterial/archaeal 16 sequences.
    &#34;&#34;&#34;

    # Constants #
    short_name = &#39;ncbi_blast&#39;
    long_name  = &#39;BLAST against the NCBI 16S database&#39;

    # The database #
    db_short_name   = &#34;ncbi_16s_rna&#34;
    db_long_name    = &#39;The NCBI 16S RNA database&#39;

    # Parameters #
    min_e_value        = 1e-5
    min_perc_identity  = 97
    max_target_seqs    = 5

    def __repr__(self):
        msg = &#39;&lt;%s object on &#34;%s&#34;&gt;&#39;
        return msg % (self.__class__.__name__, self.source.path)

    def __init__(self, source, database, dest_dir, proj):
        # Source is the FASTA file containing OTU consensus sequences #
        self.source = FASTA(source)
        # Database is an object from the `seqsearch.databases` subpackage #
        self.database = database
        # Destination is a directory that contains all the results #
        self.dest_dir = DirectoryPath(dest_dir)
        # Keep a reference to the project #
        self.proj = proj

    #----------------------------- Installing --------------------------------#
    apt_packages = [&#39;ncbi-blast+&#39;]

    def check_installed(self, exception=True):
        &#34;&#34;&#34;
        Try to determine if the BLAST software is installed and accessible.
        Also try to determine if the database chosen is downloaded and
        accessible.
        &#34;&#34;&#34;
        # The database #
        if not self.database:
            msg  = &#34;The taxonomic database does not seem to be accessible.&#34;
            msg += &#34; More information follows.\n\n&#34;
            msg += self.database.__doc__
            raise Exception(msg)
        # The command #
        return check_cmd(&#39;blastn&#39;, exception)

    #-------------------------- Automatic paths ------------------------------#
    all_paths = &#34;&#34;&#34;
                /blast/hits.xml
                /blast/stdout.txt
                /blast/stderr.txt
                /summary/single_hit.xlsx
                /summary/five_hits.xlsx
                /report.pdf
                &#34;&#34;&#34;

    @property_cached
    def autopaths(self):
        &#34;&#34;&#34;
        The AutoPaths object is used for quickly assessing the filesystem paths
        of various file inputs/outputs and directories.
        See https://github.com/xapple/autopaths#autopaths-object
        &#34;&#34;&#34;
        from autopaths.auto_paths import AutoPaths
        return AutoPaths(self.dest_dir, self.all_paths)

    #------------------------------ Running ----------------------------------#
    @property_cached
    def seq_search(self):
        &#34;&#34;&#34;
        The SeqSearch object is used for quickly launching BLAST search queries
        against a database of choice starting from a FASTA object.
        See https://github.com/xapple/seqsearch
        &#34;&#34;&#34;
        # Import #
        from seqsearch.search import SeqSearch
        # Extra options for the BLAST search #
        params = {&#39;-evalue&#39;:          self.min_e_value,
                  &#39;-perc_identity&#39;:   self.min_perc_identity,
                  &#39;-max_target_seqs&#39;: self.max_target_seqs}
        # Create the object #
        return SeqSearch(input_fasta = self.source,
                         database    = self.database,
                         seq_type    = &#39;nucl&#39;,
                         out_path    = self.autopaths.hits,
                         _out        = self.autopaths.stdout,
                         _err        = self.autopaths.stderr,
                         params      = params)

    def __call__(self, cpus=None, verbose=True):
        # Message #
        if verbose:
            message = &#34;Running `%s` taxonomy classification on &#39;%s&#39;.&#34;
            print(message % (self.short_name, self.source))
        # Check blast is installed #
        self.check_installed()
        # Run #
        self.seq_search.run()
        # Message #
        if verbose:
            message = &#34;Generating BLAST summary Excel files at &#39;%s&#39;.&#34;
            print(message % self.autopaths.summary_dir)
        # Generate summary lists #
        one  = self.all_otus_df(1)
        five = self.all_otus_df(5)
        # Create excel reports #
        one.to_excel(self.autopaths.xlsx_single_hit.path, index=False)
        five.to_excel(self.autopaths.xlsx_five_hits.path, index=False)

    #------------------------------- Results ---------------------------------#
    def __bool__(self):
        &#34;&#34;&#34;
        Return True if the taxonomic software was run already and the results
        are stored on the filesystem. Return False if it was not yet run.
        &#34;&#34;&#34;
        return self.autopaths.hits.exists

    @property_cached
    def results(self):
        # Check it was run #
        if not self:
            msg = &#34;You can&#39;t access results from the BLAST taxonomic&#34; \
                  &#34; classification before running the tool.&#34;
            raise Exception(msg)
        # Return the results #
        return BlastResults(self)

    #------------------------------ Dataframes -------------------------------#
    # Define columns needed #
    columns = [&#39;abundance&#39;, &#39;otu_id&#39;, &#39;description&#39;, &#39;score&#39;, &#39;e_value&#39;,
               &#39;accession&#39;, &#39;length&#39;, &#39;cover&#39;, &#39;identity&#39;]

    def all_otus_df(self, num_hits_per_otu=5):
        &#34;&#34;&#34;
        Returns a dataframe object summarizing all the top hits for all OTUs.
        The results looks like:

        abundance       otu_id  hit_number  ...  accession  length  identity
             4233    small_1:0           1  ...  NR_113957    1475     99.9%
             4233    small_1:0           2  ...  NR_036904    1472    100.0%
             4233    small_1:0           3  ...  NR_113405    1488     99.5%
             4233    small_1:0           4  ...  NR_118997    1524     98.6%
        &#34;&#34;&#34;
        # Parse the BLAST XML output file with BioPython #
        all_otus_hits = self.seq_search.results
        # Produce one mini-dataframe for every OTU #
        all_otus_dfs = [self.one_otu_df(record, num_hits_per_otu)
                        for record in all_otus_hits]
        # Case there are no hits in any OTUs at all #
        if all(df is None for df in all_otus_dfs):
            return pandas.DataFrame()
        # Combine them all into a big dataframe #
        df = pandas.concat(all_otus_dfs)
        # Keep the hit number as a seperate column #
        df = df.reset_index()
        col = df.pop(&#34;index&#34;)
        df.insert(2, &#39;hit_number&#39;, col)
        df[&#39;hit_number&#39;] += 1
        # Sort by abundance but keep OTUs grouped of course #
        df = df.sort_values(by=[&#39;abundance&#39;, &#39;otu_id&#39;, &#39;hit_number&#39;],
                            ascending=[False, False, True])
        # Return #
        return df

    def one_otu_df(self, record, max_hits=5):
        &#34;&#34;&#34;
        Returns a dataframe object summarizing the top hits for a given OTU.

        The record object is of type: &lt;Bio.Blast.Record.Blast&gt;

        The title looks like: &#34;gi|631252759|ref|NR_113957.1| Staphylococcus...&#34;
        The query looks like: &#34;centroid=small_1:0;seqs=4233&#34;

        To calculate the percentage identity see:

            https://github.com/peterjc/galaxy_blast/blob/master/tools/
            ncbi_blast_plus/blastxml_to_tabular.py#L230
        &#34;&#34;&#34;
        # Case there are no hits #
        if not record.descriptions: return
        # Get OTU name and its abundance #
        pattern = &#39;\Acentroid=(.+);seqs=([0-9]+)\Z&#39;
        otu, abund = re.findall(pattern, record.query)[0]
        # Cast to integer #
        abund = int(abund)
        # The description field contains lots of info we don&#39;t need #
        shorten = lambda s: s.split(&#39;|&#39;)[-1].strip()
        # Generate one line per hit #
        def one_row_per_hit():
            for align, desc in zip(record.alignments, record.descriptions):
                # Calculate all parameters #
                score       = desc.score
                e_value     = desc.e
                acc         = desc.accession
                title       = shorten(desc.title)
                hsps        = align.hsps[0]
                ident       = hsps.identities
                # The lengths available #
                qry_length  = record.query_length
                hit_length  = align.length
                algn_length = hsps.align_length
                # The ratios defined #
                p_cover = float(qry_length) / float(algn_length)
                p_ident = float(ident)      / float(algn_length)
                # Format percentages as strings #
                p_cover = &#34;%0.1f%%&#34; % (100 * p_cover)
                p_ident = &#34;%0.1f%%&#34; % (100 * p_ident)
                # Yield one row #
                yield abund, otu, title, score, e_value, acc,\
                      algn_length, p_cover, p_ident
        # Convert to dataframe #
        df = pandas.DataFrame(one_row_per_hit(), columns=self.columns)
        # Sort by E-value and also by score #
        df = df.sort_values(by=[&#39;e_value&#39;, &#39;score&#39;], ascending=False)
        # Filter and take only the top hits #
        df = df.head(max_hits)
        # Return #
        return df</code></pre>
</details>
<h3>Class variables</h3>
<dl>
<dt id="pacmill.taxonomy.ncbi_blast.BlastClassify.all_paths"><code class="name">var <span class="ident">all_paths</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="pacmill.taxonomy.ncbi_blast.BlastClassify.apt_packages"><code class="name">var <span class="ident">apt_packages</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="pacmill.taxonomy.ncbi_blast.BlastClassify.columns"><code class="name">var <span class="ident">columns</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="pacmill.taxonomy.ncbi_blast.BlastClassify.db_long_name"><code class="name">var <span class="ident">db_long_name</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="pacmill.taxonomy.ncbi_blast.BlastClassify.db_short_name"><code class="name">var <span class="ident">db_short_name</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="pacmill.taxonomy.ncbi_blast.BlastClassify.long_name"><code class="name">var <span class="ident">long_name</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="pacmill.taxonomy.ncbi_blast.BlastClassify.max_target_seqs"><code class="name">var <span class="ident">max_target_seqs</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="pacmill.taxonomy.ncbi_blast.BlastClassify.min_e_value"><code class="name">var <span class="ident">min_e_value</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="pacmill.taxonomy.ncbi_blast.BlastClassify.min_perc_identity"><code class="name">var <span class="ident">min_perc_identity</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="pacmill.taxonomy.ncbi_blast.BlastClassify.short_name"><code class="name">var <span class="ident">short_name</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
</dl>
<h3>Instance variables</h3>
<dl>
<dt id="pacmill.taxonomy.ncbi_blast.BlastClassify.autopaths"><code class="name">var <span class="ident">autopaths</span></code></dt>
<dd>
<div class="desc"><p>The AutoPaths object is used for quickly assessing the filesystem paths
of various file inputs/outputs and directories.
See <a href="https://github.com/xapple/autopaths#autopaths-object">https://github.com/xapple/autopaths#autopaths-object</a></p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def __get__(self, instance, owner):
    &#34;&#34;&#34;
    If you see the current source code in a seemingly unrelated part of
    an auto-generated documentation, it means the program making the
    documentation was unable to correctly traverse a decorated property.
    &#34;&#34;&#34;
    # For debugging purposes #
    if False: print(&#34;-&gt; property cached `%s`&#34; % instance)
    # If called from a class #
    if instance is None: return self
    # Does a cache exist for this instance? #
    self.check_cache(instance)
    # Is the answer in the cache? #
    if self.name in instance.__cache__: return instance.__cache__[self.name]
    # If not we will compute it #
    if inspect.isgeneratorfunction(self.func): result = tuple(self.func(instance))
    else:                                      result = self.func(instance)
    # Let&#39;s store the answer for later #
    instance.__cache__[self.name] = result
    # Return #
    return result</code></pre>
</details>
</dd>
<dt id="pacmill.taxonomy.ncbi_blast.BlastClassify.results"><code class="name">var <span class="ident">results</span></code></dt>
<dd>
<div class="desc"><p>If you see the current source code in a seemingly unrelated part of
an auto-generated documentation, it means the program making the
documentation was unable to correctly traverse a decorated property.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def __get__(self, instance, owner):
    &#34;&#34;&#34;
    If you see the current source code in a seemingly unrelated part of
    an auto-generated documentation, it means the program making the
    documentation was unable to correctly traverse a decorated property.
    &#34;&#34;&#34;
    # For debugging purposes #
    if False: print(&#34;-&gt; property cached `%s`&#34; % instance)
    # If called from a class #
    if instance is None: return self
    # Does a cache exist for this instance? #
    self.check_cache(instance)
    # Is the answer in the cache? #
    if self.name in instance.__cache__: return instance.__cache__[self.name]
    # If not we will compute it #
    if inspect.isgeneratorfunction(self.func): result = tuple(self.func(instance))
    else:                                      result = self.func(instance)
    # Let&#39;s store the answer for later #
    instance.__cache__[self.name] = result
    # Return #
    return result</code></pre>
</details>
</dd>
<dt id="pacmill.taxonomy.ncbi_blast.BlastClassify.seq_search"><code class="name">var <span class="ident">seq_search</span></code></dt>
<dd>
<div class="desc"><p>The SeqSearch object is used for quickly launching BLAST search queries
against a database of choice starting from a FASTA object.
See <a href="https://github.com/xapple/seqsearch">https://github.com/xapple/seqsearch</a></p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def __get__(self, instance, owner):
    &#34;&#34;&#34;
    If you see the current source code in a seemingly unrelated part of
    an auto-generated documentation, it means the program making the
    documentation was unable to correctly traverse a decorated property.
    &#34;&#34;&#34;
    # For debugging purposes #
    if False: print(&#34;-&gt; property cached `%s`&#34; % instance)
    # If called from a class #
    if instance is None: return self
    # Does a cache exist for this instance? #
    self.check_cache(instance)
    # Is the answer in the cache? #
    if self.name in instance.__cache__: return instance.__cache__[self.name]
    # If not we will compute it #
    if inspect.isgeneratorfunction(self.func): result = tuple(self.func(instance))
    else:                                      result = self.func(instance)
    # Let&#39;s store the answer for later #
    instance.__cache__[self.name] = result
    # Return #
    return result</code></pre>
</details>
</dd>
</dl>
<h3>Methods</h3>
<dl>
<dt id="pacmill.taxonomy.ncbi_blast.BlastClassify.all_otus_df"><code class="name flex">
<span>def <span class="ident">all_otus_df</span></span>(<span>self, num_hits_per_otu=5)</span>
</code></dt>
<dd>
<div class="desc"><p>Returns a dataframe object summarizing all the top hits for all OTUs.
The results looks like:</p>
<p>abundance
otu_id
hit_number
&hellip;
accession
length
identity
4233
small_1:0
1
&hellip;
NR_113957
1475
99.9%
4233
small_1:0
2
&hellip;
NR_036904
1472
100.0%
4233
small_1:0
3
&hellip;
NR_113405
1488
99.5%
4233
small_1:0
4
&hellip;
NR_118997
1524
98.6%</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def all_otus_df(self, num_hits_per_otu=5):
    &#34;&#34;&#34;
    Returns a dataframe object summarizing all the top hits for all OTUs.
    The results looks like:

    abundance       otu_id  hit_number  ...  accession  length  identity
         4233    small_1:0           1  ...  NR_113957    1475     99.9%
         4233    small_1:0           2  ...  NR_036904    1472    100.0%
         4233    small_1:0           3  ...  NR_113405    1488     99.5%
         4233    small_1:0           4  ...  NR_118997    1524     98.6%
    &#34;&#34;&#34;
    # Parse the BLAST XML output file with BioPython #
    all_otus_hits = self.seq_search.results
    # Produce one mini-dataframe for every OTU #
    all_otus_dfs = [self.one_otu_df(record, num_hits_per_otu)
                    for record in all_otus_hits]
    # Case there are no hits in any OTUs at all #
    if all(df is None for df in all_otus_dfs):
        return pandas.DataFrame()
    # Combine them all into a big dataframe #
    df = pandas.concat(all_otus_dfs)
    # Keep the hit number as a seperate column #
    df = df.reset_index()
    col = df.pop(&#34;index&#34;)
    df.insert(2, &#39;hit_number&#39;, col)
    df[&#39;hit_number&#39;] += 1
    # Sort by abundance but keep OTUs grouped of course #
    df = df.sort_values(by=[&#39;abundance&#39;, &#39;otu_id&#39;, &#39;hit_number&#39;],
                        ascending=[False, False, True])
    # Return #
    return df</code></pre>
</details>
</dd>
<dt id="pacmill.taxonomy.ncbi_blast.BlastClassify.check_installed"><code class="name flex">
<span>def <span class="ident">check_installed</span></span>(<span>self, exception=True)</span>
</code></dt>
<dd>
<div class="desc"><p>Try to determine if the BLAST software is installed and accessible.
Also try to determine if the database chosen is downloaded and
accessible.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def check_installed(self, exception=True):
    &#34;&#34;&#34;
    Try to determine if the BLAST software is installed and accessible.
    Also try to determine if the database chosen is downloaded and
    accessible.
    &#34;&#34;&#34;
    # The database #
    if not self.database:
        msg  = &#34;The taxonomic database does not seem to be accessible.&#34;
        msg += &#34; More information follows.\n\n&#34;
        msg += self.database.__doc__
        raise Exception(msg)
    # The command #
    return check_cmd(&#39;blastn&#39;, exception)</code></pre>
</details>
</dd>
<dt id="pacmill.taxonomy.ncbi_blast.BlastClassify.one_otu_df"><code class="name flex">
<span>def <span class="ident">one_otu_df</span></span>(<span>self, record, max_hits=5)</span>
</code></dt>
<dd>
<div class="desc"><p>Returns a dataframe object summarizing the top hits for a given OTU.</p>
<p>The record object is of type: <Bio.Blast.Record.Blast></p>
<p>The title looks like: "gi|631252759|ref|NR_113957.1| Staphylococcus&hellip;"
The query looks like: "centroid=small_1:0;seqs=4233"</p>
<p>To calculate the percentage identity see:</p>
<pre><code>&lt;https://github.com/peterjc/galaxy_blast/blob/master/tools/&gt;
ncbi_blast_plus/blastxml_to_tabular.py#L230
</code></pre></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def one_otu_df(self, record, max_hits=5):
    &#34;&#34;&#34;
    Returns a dataframe object summarizing the top hits for a given OTU.

    The record object is of type: &lt;Bio.Blast.Record.Blast&gt;

    The title looks like: &#34;gi|631252759|ref|NR_113957.1| Staphylococcus...&#34;
    The query looks like: &#34;centroid=small_1:0;seqs=4233&#34;

    To calculate the percentage identity see:

        https://github.com/peterjc/galaxy_blast/blob/master/tools/
        ncbi_blast_plus/blastxml_to_tabular.py#L230
    &#34;&#34;&#34;
    # Case there are no hits #
    if not record.descriptions: return
    # Get OTU name and its abundance #
    pattern = &#39;\Acentroid=(.+);seqs=([0-9]+)\Z&#39;
    otu, abund = re.findall(pattern, record.query)[0]
    # Cast to integer #
    abund = int(abund)
    # The description field contains lots of info we don&#39;t need #
    shorten = lambda s: s.split(&#39;|&#39;)[-1].strip()
    # Generate one line per hit #
    def one_row_per_hit():
        for align, desc in zip(record.alignments, record.descriptions):
            # Calculate all parameters #
            score       = desc.score
            e_value     = desc.e
            acc         = desc.accession
            title       = shorten(desc.title)
            hsps        = align.hsps[0]
            ident       = hsps.identities
            # The lengths available #
            qry_length  = record.query_length
            hit_length  = align.length
            algn_length = hsps.align_length
            # The ratios defined #
            p_cover = float(qry_length) / float(algn_length)
            p_ident = float(ident)      / float(algn_length)
            # Format percentages as strings #
            p_cover = &#34;%0.1f%%&#34; % (100 * p_cover)
            p_ident = &#34;%0.1f%%&#34; % (100 * p_ident)
            # Yield one row #
            yield abund, otu, title, score, e_value, acc,\
                  algn_length, p_cover, p_ident
    # Convert to dataframe #
    df = pandas.DataFrame(one_row_per_hit(), columns=self.columns)
    # Sort by E-value and also by score #
    df = df.sort_values(by=[&#39;e_value&#39;, &#39;score&#39;], ascending=False)
    # Filter and take only the top hits #
    df = df.head(max_hits)
    # Return #
    return df</code></pre>
</details>
</dd>
</dl>
</dd>
<dt id="pacmill.taxonomy.ncbi_blast.BlastResults"><code class="flex name class">
<span>class <span class="ident">BlastResults</span></span>
<span>(</span><span>tax)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class BlastResults:

    def __init__(self, tax):
        self.tax = tax

    @property_cached
    def report(self):
        &#34;&#34;&#34;
        The NcbiBlastReport object is used to produce a PDF document containing
        all the information that concern this particular taxonomic
        classification strategy.
        &#34;&#34;&#34;
        # Imports #
        from pacmill.reports.ncbi_blast import NcbiBlastReport
        # Output directory #
        path = self.tax.proj.autopaths.taxonomy_dir + &#39;reports/&#39;
        path = path + self.tax.short_name + &#39;.pdf&#39;
        # Return #
        return NcbiBlastReport(self.tax, path)</code></pre>
</details>
<h3>Instance variables</h3>
<dl>
<dt id="pacmill.taxonomy.ncbi_blast.BlastResults.report"><code class="name">var <span class="ident">report</span></code></dt>
<dd>
<div class="desc"><p>The NcbiBlastReport object is used to produce a PDF document containing
all the information that concern this particular taxonomic
classification strategy.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def __get__(self, instance, owner):
    &#34;&#34;&#34;
    If you see the current source code in a seemingly unrelated part of
    an auto-generated documentation, it means the program making the
    documentation was unable to correctly traverse a decorated property.
    &#34;&#34;&#34;
    # For debugging purposes #
    if False: print(&#34;-&gt; property cached `%s`&#34; % instance)
    # If called from a class #
    if instance is None: return self
    # Does a cache exist for this instance? #
    self.check_cache(instance)
    # Is the answer in the cache? #
    if self.name in instance.__cache__: return instance.__cache__[self.name]
    # If not we will compute it #
    if inspect.isgeneratorfunction(self.func): result = tuple(self.func(instance))
    else:                                      result = self.func(instance)
    # Let&#39;s store the answer for later #
    instance.__cache__[self.name] = result
    # Return #
    return result</code></pre>
</details>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="pacmill.taxonomy" href="index.html">pacmill.taxonomy</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="pacmill.taxonomy.ncbi_blast.BlastClassify" href="#pacmill.taxonomy.ncbi_blast.BlastClassify">BlastClassify</a></code></h4>
<ul class="two-column">
<li><code><a title="pacmill.taxonomy.ncbi_blast.BlastClassify.all_otus_df" href="#pacmill.taxonomy.ncbi_blast.BlastClassify.all_otus_df">all_otus_df</a></code></li>
<li><code><a title="pacmill.taxonomy.ncbi_blast.BlastClassify.all_paths" href="#pacmill.taxonomy.ncbi_blast.BlastClassify.all_paths">all_paths</a></code></li>
<li><code><a title="pacmill.taxonomy.ncbi_blast.BlastClassify.apt_packages" href="#pacmill.taxonomy.ncbi_blast.BlastClassify.apt_packages">apt_packages</a></code></li>
<li><code><a title="pacmill.taxonomy.ncbi_blast.BlastClassify.autopaths" href="#pacmill.taxonomy.ncbi_blast.BlastClassify.autopaths">autopaths</a></code></li>
<li><code><a title="pacmill.taxonomy.ncbi_blast.BlastClassify.check_installed" href="#pacmill.taxonomy.ncbi_blast.BlastClassify.check_installed">check_installed</a></code></li>
<li><code><a title="pacmill.taxonomy.ncbi_blast.BlastClassify.columns" href="#pacmill.taxonomy.ncbi_blast.BlastClassify.columns">columns</a></code></li>
<li><code><a title="pacmill.taxonomy.ncbi_blast.BlastClassify.db_long_name" href="#pacmill.taxonomy.ncbi_blast.BlastClassify.db_long_name">db_long_name</a></code></li>
<li><code><a title="pacmill.taxonomy.ncbi_blast.BlastClassify.db_short_name" href="#pacmill.taxonomy.ncbi_blast.BlastClassify.db_short_name">db_short_name</a></code></li>
<li><code><a title="pacmill.taxonomy.ncbi_blast.BlastClassify.long_name" href="#pacmill.taxonomy.ncbi_blast.BlastClassify.long_name">long_name</a></code></li>
<li><code><a title="pacmill.taxonomy.ncbi_blast.BlastClassify.max_target_seqs" href="#pacmill.taxonomy.ncbi_blast.BlastClassify.max_target_seqs">max_target_seqs</a></code></li>
<li><code><a title="pacmill.taxonomy.ncbi_blast.BlastClassify.min_e_value" href="#pacmill.taxonomy.ncbi_blast.BlastClassify.min_e_value">min_e_value</a></code></li>
<li><code><a title="pacmill.taxonomy.ncbi_blast.BlastClassify.min_perc_identity" href="#pacmill.taxonomy.ncbi_blast.BlastClassify.min_perc_identity">min_perc_identity</a></code></li>
<li><code><a title="pacmill.taxonomy.ncbi_blast.BlastClassify.one_otu_df" href="#pacmill.taxonomy.ncbi_blast.BlastClassify.one_otu_df">one_otu_df</a></code></li>
<li><code><a title="pacmill.taxonomy.ncbi_blast.BlastClassify.results" href="#pacmill.taxonomy.ncbi_blast.BlastClassify.results">results</a></code></li>
<li><code><a title="pacmill.taxonomy.ncbi_blast.BlastClassify.seq_search" href="#pacmill.taxonomy.ncbi_blast.BlastClassify.seq_search">seq_search</a></code></li>
<li><code><a title="pacmill.taxonomy.ncbi_blast.BlastClassify.short_name" href="#pacmill.taxonomy.ncbi_blast.BlastClassify.short_name">short_name</a></code></li>
</ul>
</li>
<li>
<h4><code><a title="pacmill.taxonomy.ncbi_blast.BlastResults" href="#pacmill.taxonomy.ncbi_blast.BlastResults">BlastResults</a></code></h4>
<ul class="">
<li><code><a title="pacmill.taxonomy.ncbi_blast.BlastResults.report" href="#pacmill.taxonomy.ncbi_blast.BlastResults.report">report</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.9.2</a>.</p>
</footer>
</body>
</html>