<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.8.4" />
<title>pacmill.filtering.barrnap API documentation</title>
<meta name="description" content="Written by Lucas Sinclair.
MIT Licensed.
Contact at www.sinclair.bio" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.18.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.18.1/highlight.min.js" integrity="sha256-eOgo0OtLL4cdq7RdwRUiGKLX9XsIJ7nGhWEKbohmVAQ=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>pacmill.filtering.barrnap</code></h1>
</header>
<section id="section-intro">
<p>Written by Lucas Sinclair.
MIT Licensed.
Contact at www.sinclair.bio</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">#!/usr/bin/env python3
# -*- coding: utf-8 -*-

&#34;&#34;&#34;
Written by Lucas Sinclair.
MIT Licensed.
Contact at www.sinclair.bio
&#34;&#34;&#34;

# Built-in modules #
import os, multiprocessing

# First party modules #
from fasta import FASTA, FASTQ
from autopaths.file_path      import FilePath
from autopaths.tmp_path       import new_temp_dir, new_temp_path
from plumbing.cache           import property_cached
from plumbing.check_cmd_found import check_cmd
from plumbing.apt_pkg         import get_apt_packages
from plumbing.scraping        import download_from_url

# Third party modules #
import sh, tag

###############################################################################
class Barrnap:
    &#34;&#34;&#34;
    Takes care of running the Barrnap program on a given FASTQ file.
    See https://github.com/tseemann/barrnap
    Barrnap predicts the presence and location of ribosomal RNA genes in
    genomes. Expects version 0.9.

    The output of the program is a GFF3 and looks something like:

        P.marinus  barrnap:0.9  rRNA  3314  4793  0       +  .  Name=16S_rRNA
        P.marinus  barrnap:0.9  rRNA  5464  8334  0       +  .  Name=23S_rRNA
        P.marinus  barrnap:0.9  rRNA  8433  8536  9.6e-07 +  .  Name=5S_rRNA

    You can either use barrnap in filter mode, or in extract mode.
    See the two child classes below.
    &#34;&#34;&#34;

    # Proportional length threshold to reject prediction #
    reject_threshold = 0.4

    def __repr__(self):
        msg = &#39;&lt;%s object on &#34;%s&#34;&gt;&#39;
        return msg % (self.__class__.__name__, self.source.path)

    def __init__(self, source, dest=None, filtered=None):
        # Source is the FASTA or FASTQ file on which barrnap will be run #
        if source.endswith(&#39;fasta&#39;): self.source = FASTA(source)
        else:                        self.source = FASTQ(source)
        # Destination is a GFF file that contains the results #
        if dest is None:
            dest = self.source.prefix_path + &#39;.barrnap.gff&#39;
        self.dest = FilePath(dest)
        # Filtered is a file containing only raw reads with rRNA genes #
        if filtered is None:
            filtered = self.source.prefix_path + &#39;.barrnap.fastq&#39;
        # It can be either a FASTA or a FASTQ #
        if filtered.endswith(&#39;fasta&#39;): self.filtered = FASTA(filtered)
        else:                          self.filtered = FASTQ(filtered)

    #---------------------------- Installing ---------------------------------#
    apt_packages = [&#39;bedtools&#39;]
    zip_url = &#34;https://github.com/tseemann/barrnap/archive/0.9.zip&#34;

    @classmethod
    def check_installed(cls, exception=True):
        &#34;&#34;&#34;
        Try to determine if the Barrnap software is installed and
        accessible.
        &#34;&#34;&#34;
        return check_cmd(&#39;barrnap&#39;, exception, cls.install.__doc__)

    @classmethod
    def install(cls, prefix=&#34;~/programs/barrnap/&#34;):
        &#34;&#34;&#34;
        To automatically download and install the Barrnap software on this
        computer and for the current user, type these commands in python:

            &gt;&gt;&gt; from pacmill.filtering.barrnap import Barrnap
            &gt;&gt;&gt; Barrnap.install()
        &#34;&#34;&#34;
        # Start with required apt packages #
        get_apt_packages(cls.apt_packages, verbose=True)
        # Make a temporary directory #
        tmp_dir = new_temp_dir()
        # Download tarball #
        zip_loc = download_from_url(cls.zip_url, tmp_dir, stream=True,
                                    progress=True)
        # Uncompress #
        zip_loc.unzip_to(tmp_dir, single=False)
        src_dir = tmp_dir.sub_directory
        src_dir.move_to(prefix)
        # Set executable permissions #
        bin_loc = src_dir + &#39;bin/barrnap&#39;
        bin_loc.permissions.make_executable()
        bin_loc = src_dir + &#39;binaries/linux/nhmmer&#39;
        bin_loc.permissions.make_executable()
        # The directory that contains the executable #
        bin_dir = src_dir.with_tilda[:-1].replace(&#39;~&#39;, &#39;$HOME&#39;)
        # Suggest adding to the $PATH #
        print(&#34;\nBarrnap was installed successfully. You should now &#34;
              &#34;add this line to your .bash_profile: \n\n    &#34;
              &#34;export PATH=%s/bin:$PATH\n&#34; % bin_dir)

    #------------------------------ Running ----------------------------------#
    def run(self, cpus=None, verbose=True):
        # Message #
        if verbose: print(&#34;Running barrnap on &#39;%s&#39;&#34; % self.source)
        # Check it is installed #
        self.check_installed()
        # Check version #
        assert b&#34;0.9&#34; in sh.barrnap(&#39;--version&#39;).stderr
        # If the input is a FASTQ we need to make a FASTA first #
        if self.source.endswith(&#39;fastq&#39;):
            source = new_temp_path(suffix=&#39;.fasta&#39;)
            FASTQ(self.source).to_fasta(source)
        else: source = self.source
        # Number of cores #
        if cpus is None: cpus = min(multiprocessing.cpu_count(), 32)
        # Run it #
        sh.barrnap(source, &#39;--threads&#39;, cpus, &#39;--reject&#39;,
                   self.reject_threshold, _out=str(self.dest))
        # Return #
        return self.dest

    #------------------------------- Results ---------------------------------#
    def __bool__(self):
        &#34;&#34;&#34;
        Return True if the Barrnap software was run already and the results are
        stored on the filesystem. Return False if it was not yet run.
        &#34;&#34;&#34;
        return self.filtered.exists

    @property_cached
    def results(self):
        # Check it was run #
        if not self:
            msg = &#34;You can&#39;t access results from Barrnap &#34; \
                  &#34;before running the tool.&#34;
            raise Exception(msg)
        # Return the results #
        return self.filtered

###############################################################################
class BarrnapFilter(Barrnap):
    &#34;&#34;&#34;
    With this subclass, we will filter reads after running barrnap.

    Using the original reads file and the GFF output of barrnap,
    we will create a new FASTQ file containing only the original reads
    that had a hit for both rRNA genes on the same read.
    &#34;&#34;&#34;

    def __call__(self, cpus=None, verbose=True):
        # Call the parent class method #
        self.run(cpus, verbose)
        # Filter #
        self.filter()

    #--------------------------- Presence of gene ----------------------------#
    def parse_hit_ids(self, attr_text):
        &#34;&#34;&#34;
        Using the GFF output of barrnap and the `tag` parser, we retrieve
        all IDs that contained a given text in their attributes.
        &#34;&#34;&#34;
        reader = tag.GFF3Reader(infilename=self.dest)
        reader = tag.select.features(reader, type=&#39;rRNA&#39;)
        return [rec.seqid for rec in reader if attr_text in rec.attributes]

    @property_cached
    def ids_16s(self):
        &#34;&#34;&#34;Return read IDs for reads that contained a 16S gene.&#34;&#34;&#34;
        return frozenset(self.parse_hit_ids(&#39;16S_rRNA&#39;))

    @property_cached
    def ids_23s(self):
        &#34;&#34;&#34;Return read IDs for reads that contained a 23S gene.&#34;&#34;&#34;
        return frozenset(self.parse_hit_ids(&#39;23S_rRNA&#39;))

    def filter(self, verbose=True):
        # Message #
        if verbose:
            msg = &#34;Extracting sequences with both rRNA genes from &#39;%s&#39;&#34;
            print(msg % self.dest)
        # Get those that had both genes #
        ids_both = self.ids_16s &amp; self.ids_23s
        # Extract those IDs #
        FASTQ(self.source).extract_sequences(ids_both, self.filtered)
        # Return #
        return self.filtered

###############################################################################
class BarrnapExtract(Barrnap):
    &#34;&#34;&#34;
    With this subclass, we will extract reads regions after running barrnap.

    Using the original reads file and the GFF output of
    barrnap, we will create a new FASTA file containing only the portion of
    the original reads that contains the 16S rRNA gene.
    &#34;&#34;&#34;

    def __call__(self, cpus=None, verbose=True):
        # Call the parent class method #
        self.run(cpus, verbose)
        # Filter #
        self.extract()

    #--------------------------- Location of gene ----------------------------#
    def extract(self, verbose=True):
        # Message #
        if verbose:
            msg = &#34;Extracting the 16S rRNA portion of sequences from &#39;%s&#39;&#34;
            print(msg % self.dest)
        # Get reads that had a rRNA hit #
        reader = tag.GFF3Reader(infilename=self.dest)
        reader = tag.select.features(reader, type=&#39;rRNA&#39;)
        # Make a lookup table for those with 16S hits #
        lookup = {rec.seqid: (rec.start, rec.end)
                  for rec in reader if &#39;16S_rRNA&#39; in rec.attributes}
        # Function to yield only the good part of each read #
        def only_16s_portion(reads):
            for r in reads:
                start_and_end = lookup.get(r.id)
                if start_and_end is None: continue
                start, end = start_and_end
                yield r[start:end]
        # Write new FASTA file #
        self.filtered.write(only_16s_portion(self.source))
        # Return #
        return self.filtered</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="pacmill.filtering.barrnap.Barrnap"><code class="flex name class">
<span>class <span class="ident">Barrnap</span></span>
<span>(</span><span>source, dest=None, filtered=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Takes care of running the Barrnap program on a given FASTQ file.
See <a href="https://github.com/tseemann/barrnap">https://github.com/tseemann/barrnap</a>
Barrnap predicts the presence and location of ribosomal RNA genes in
genomes. Expects version 0.9.</p>
<p>The output of the program is a GFF3 and looks something like:</p>
<pre><code>P.marinus  barrnap:0.9  rRNA  3314  4793  0       +  .  Name=16S_rRNA
P.marinus  barrnap:0.9  rRNA  5464  8334  0       +  .  Name=23S_rRNA
P.marinus  barrnap:0.9  rRNA  8433  8536  9.6e-07 +  .  Name=5S_rRNA
</code></pre>
<p>You can either use barrnap in filter mode, or in extract mode.
See the two child classes below.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class Barrnap:
    &#34;&#34;&#34;
    Takes care of running the Barrnap program on a given FASTQ file.
    See https://github.com/tseemann/barrnap
    Barrnap predicts the presence and location of ribosomal RNA genes in
    genomes. Expects version 0.9.

    The output of the program is a GFF3 and looks something like:

        P.marinus  barrnap:0.9  rRNA  3314  4793  0       +  .  Name=16S_rRNA
        P.marinus  barrnap:0.9  rRNA  5464  8334  0       +  .  Name=23S_rRNA
        P.marinus  barrnap:0.9  rRNA  8433  8536  9.6e-07 +  .  Name=5S_rRNA

    You can either use barrnap in filter mode, or in extract mode.
    See the two child classes below.
    &#34;&#34;&#34;

    # Proportional length threshold to reject prediction #
    reject_threshold = 0.4

    def __repr__(self):
        msg = &#39;&lt;%s object on &#34;%s&#34;&gt;&#39;
        return msg % (self.__class__.__name__, self.source.path)

    def __init__(self, source, dest=None, filtered=None):
        # Source is the FASTA or FASTQ file on which barrnap will be run #
        if source.endswith(&#39;fasta&#39;): self.source = FASTA(source)
        else:                        self.source = FASTQ(source)
        # Destination is a GFF file that contains the results #
        if dest is None:
            dest = self.source.prefix_path + &#39;.barrnap.gff&#39;
        self.dest = FilePath(dest)
        # Filtered is a file containing only raw reads with rRNA genes #
        if filtered is None:
            filtered = self.source.prefix_path + &#39;.barrnap.fastq&#39;
        # It can be either a FASTA or a FASTQ #
        if filtered.endswith(&#39;fasta&#39;): self.filtered = FASTA(filtered)
        else:                          self.filtered = FASTQ(filtered)

    #---------------------------- Installing ---------------------------------#
    apt_packages = [&#39;bedtools&#39;]
    zip_url = &#34;https://github.com/tseemann/barrnap/archive/0.9.zip&#34;

    @classmethod
    def check_installed(cls, exception=True):
        &#34;&#34;&#34;
        Try to determine if the Barrnap software is installed and
        accessible.
        &#34;&#34;&#34;
        return check_cmd(&#39;barrnap&#39;, exception, cls.install.__doc__)

    @classmethod
    def install(cls, prefix=&#34;~/programs/barrnap/&#34;):
        &#34;&#34;&#34;
        To automatically download and install the Barrnap software on this
        computer and for the current user, type these commands in python:

            &gt;&gt;&gt; from pacmill.filtering.barrnap import Barrnap
            &gt;&gt;&gt; Barrnap.install()
        &#34;&#34;&#34;
        # Start with required apt packages #
        get_apt_packages(cls.apt_packages, verbose=True)
        # Make a temporary directory #
        tmp_dir = new_temp_dir()
        # Download tarball #
        zip_loc = download_from_url(cls.zip_url, tmp_dir, stream=True,
                                    progress=True)
        # Uncompress #
        zip_loc.unzip_to(tmp_dir, single=False)
        src_dir = tmp_dir.sub_directory
        src_dir.move_to(prefix)
        # Set executable permissions #
        bin_loc = src_dir + &#39;bin/barrnap&#39;
        bin_loc.permissions.make_executable()
        bin_loc = src_dir + &#39;binaries/linux/nhmmer&#39;
        bin_loc.permissions.make_executable()
        # The directory that contains the executable #
        bin_dir = src_dir.with_tilda[:-1].replace(&#39;~&#39;, &#39;$HOME&#39;)
        # Suggest adding to the $PATH #
        print(&#34;\nBarrnap was installed successfully. You should now &#34;
              &#34;add this line to your .bash_profile: \n\n    &#34;
              &#34;export PATH=%s/bin:$PATH\n&#34; % bin_dir)

    #------------------------------ Running ----------------------------------#
    def run(self, cpus=None, verbose=True):
        # Message #
        if verbose: print(&#34;Running barrnap on &#39;%s&#39;&#34; % self.source)
        # Check it is installed #
        self.check_installed()
        # Check version #
        assert b&#34;0.9&#34; in sh.barrnap(&#39;--version&#39;).stderr
        # If the input is a FASTQ we need to make a FASTA first #
        if self.source.endswith(&#39;fastq&#39;):
            source = new_temp_path(suffix=&#39;.fasta&#39;)
            FASTQ(self.source).to_fasta(source)
        else: source = self.source
        # Number of cores #
        if cpus is None: cpus = min(multiprocessing.cpu_count(), 32)
        # Run it #
        sh.barrnap(source, &#39;--threads&#39;, cpus, &#39;--reject&#39;,
                   self.reject_threshold, _out=str(self.dest))
        # Return #
        return self.dest

    #------------------------------- Results ---------------------------------#
    def __bool__(self):
        &#34;&#34;&#34;
        Return True if the Barrnap software was run already and the results are
        stored on the filesystem. Return False if it was not yet run.
        &#34;&#34;&#34;
        return self.filtered.exists

    @property_cached
    def results(self):
        # Check it was run #
        if not self:
            msg = &#34;You can&#39;t access results from Barrnap &#34; \
                  &#34;before running the tool.&#34;
            raise Exception(msg)
        # Return the results #
        return self.filtered</code></pre>
</details>
<h3>Subclasses</h3>
<ul class="hlist">
<li><a title="pacmill.filtering.barrnap.BarrnapExtract" href="#pacmill.filtering.barrnap.BarrnapExtract">BarrnapExtract</a></li>
<li><a title="pacmill.filtering.barrnap.BarrnapFilter" href="#pacmill.filtering.barrnap.BarrnapFilter">BarrnapFilter</a></li>
</ul>
<h3>Class variables</h3>
<dl>
<dt id="pacmill.filtering.barrnap.Barrnap.apt_packages"><code class="name">var <span class="ident">apt_packages</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="pacmill.filtering.barrnap.Barrnap.reject_threshold"><code class="name">var <span class="ident">reject_threshold</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
<dt id="pacmill.filtering.barrnap.Barrnap.zip_url"><code class="name">var <span class="ident">zip_url</span></code></dt>
<dd>
<div class="desc"></div>
</dd>
</dl>
<h3>Static methods</h3>
<dl>
<dt id="pacmill.filtering.barrnap.Barrnap.check_installed"><code class="name flex">
<span>def <span class="ident">check_installed</span></span>(<span>exception=True)</span>
</code></dt>
<dd>
<div class="desc"><p>Try to determine if the Barrnap software is installed and
accessible.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@classmethod
def check_installed(cls, exception=True):
    &#34;&#34;&#34;
    Try to determine if the Barrnap software is installed and
    accessible.
    &#34;&#34;&#34;
    return check_cmd(&#39;barrnap&#39;, exception, cls.install.__doc__)</code></pre>
</details>
</dd>
<dt id="pacmill.filtering.barrnap.Barrnap.install"><code class="name flex">
<span>def <span class="ident">install</span></span>(<span>prefix='~/programs/barrnap/')</span>
</code></dt>
<dd>
<div class="desc"><p>To automatically download and install the Barrnap software on this
computer and for the current user, type these commands in python:</p>
<pre><code>&gt;&gt;&gt; from pacmill.filtering.barrnap import Barrnap
&gt;&gt;&gt; Barrnap.install()
</code></pre></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@classmethod
def install(cls, prefix=&#34;~/programs/barrnap/&#34;):
    &#34;&#34;&#34;
    To automatically download and install the Barrnap software on this
    computer and for the current user, type these commands in python:

        &gt;&gt;&gt; from pacmill.filtering.barrnap import Barrnap
        &gt;&gt;&gt; Barrnap.install()
    &#34;&#34;&#34;
    # Start with required apt packages #
    get_apt_packages(cls.apt_packages, verbose=True)
    # Make a temporary directory #
    tmp_dir = new_temp_dir()
    # Download tarball #
    zip_loc = download_from_url(cls.zip_url, tmp_dir, stream=True,
                                progress=True)
    # Uncompress #
    zip_loc.unzip_to(tmp_dir, single=False)
    src_dir = tmp_dir.sub_directory
    src_dir.move_to(prefix)
    # Set executable permissions #
    bin_loc = src_dir + &#39;bin/barrnap&#39;
    bin_loc.permissions.make_executable()
    bin_loc = src_dir + &#39;binaries/linux/nhmmer&#39;
    bin_loc.permissions.make_executable()
    # The directory that contains the executable #
    bin_dir = src_dir.with_tilda[:-1].replace(&#39;~&#39;, &#39;$HOME&#39;)
    # Suggest adding to the $PATH #
    print(&#34;\nBarrnap was installed successfully. You should now &#34;
          &#34;add this line to your .bash_profile: \n\n    &#34;
          &#34;export PATH=%s/bin:$PATH\n&#34; % bin_dir)</code></pre>
</details>
</dd>
</dl>
<h3>Instance variables</h3>
<dl>
<dt id="pacmill.filtering.barrnap.Barrnap.results"><code class="name">var <span class="ident">results</span></code></dt>
<dd>
<div class="desc"><p>If you see the current source code in a seemingly unrelated part of
an auto-generated documentation, it means the program making the
documentation was unable to correctly traverse a decorated property.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def __get__(self, instance, owner):
    &#34;&#34;&#34;
    If you see the current source code in a seemingly unrelated part of
    an auto-generated documentation, it means the program making the
    documentation was unable to correctly traverse a decorated property.
    &#34;&#34;&#34;
    # For debugging purposes #
    if False: print(&#34;-&gt; property cached `%s`&#34; % instance)
    # If called from a class #
    if instance is None: return self
    # Does a cache exist for this instance? #
    self.check_cache(instance)
    # Is the answer in the cache? #
    if self.name in instance.__cache__: return instance.__cache__[self.name]
    # If not we will compute it #
    if inspect.isgeneratorfunction(self.func): result = tuple(self.func(instance))
    else:                                      result = self.func(instance)
    # Let&#39;s store the answer for later #
    instance.__cache__[self.name] = result
    # Return #
    return result</code></pre>
</details>
</dd>
</dl>
<h3>Methods</h3>
<dl>
<dt id="pacmill.filtering.barrnap.Barrnap.run"><code class="name flex">
<span>def <span class="ident">run</span></span>(<span>self, cpus=None, verbose=True)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def run(self, cpus=None, verbose=True):
    # Message #
    if verbose: print(&#34;Running barrnap on &#39;%s&#39;&#34; % self.source)
    # Check it is installed #
    self.check_installed()
    # Check version #
    assert b&#34;0.9&#34; in sh.barrnap(&#39;--version&#39;).stderr
    # If the input is a FASTQ we need to make a FASTA first #
    if self.source.endswith(&#39;fastq&#39;):
        source = new_temp_path(suffix=&#39;.fasta&#39;)
        FASTQ(self.source).to_fasta(source)
    else: source = self.source
    # Number of cores #
    if cpus is None: cpus = min(multiprocessing.cpu_count(), 32)
    # Run it #
    sh.barrnap(source, &#39;--threads&#39;, cpus, &#39;--reject&#39;,
               self.reject_threshold, _out=str(self.dest))
    # Return #
    return self.dest</code></pre>
</details>
</dd>
</dl>
</dd>
<dt id="pacmill.filtering.barrnap.BarrnapExtract"><code class="flex name class">
<span>class <span class="ident">BarrnapExtract</span></span>
<span>(</span><span>source, dest=None, filtered=None)</span>
</code></dt>
<dd>
<div class="desc"><p>With this subclass, we will extract reads regions after running barrnap.</p>
<p>Using the original reads file and the GFF output of
barrnap, we will create a new FASTA file containing only the portion of
the original reads that contains the 16S rRNA gene.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class BarrnapExtract(Barrnap):
    &#34;&#34;&#34;
    With this subclass, we will extract reads regions after running barrnap.

    Using the original reads file and the GFF output of
    barrnap, we will create a new FASTA file containing only the portion of
    the original reads that contains the 16S rRNA gene.
    &#34;&#34;&#34;

    def __call__(self, cpus=None, verbose=True):
        # Call the parent class method #
        self.run(cpus, verbose)
        # Filter #
        self.extract()

    #--------------------------- Location of gene ----------------------------#
    def extract(self, verbose=True):
        # Message #
        if verbose:
            msg = &#34;Extracting the 16S rRNA portion of sequences from &#39;%s&#39;&#34;
            print(msg % self.dest)
        # Get reads that had a rRNA hit #
        reader = tag.GFF3Reader(infilename=self.dest)
        reader = tag.select.features(reader, type=&#39;rRNA&#39;)
        # Make a lookup table for those with 16S hits #
        lookup = {rec.seqid: (rec.start, rec.end)
                  for rec in reader if &#39;16S_rRNA&#39; in rec.attributes}
        # Function to yield only the good part of each read #
        def only_16s_portion(reads):
            for r in reads:
                start_and_end = lookup.get(r.id)
                if start_and_end is None: continue
                start, end = start_and_end
                yield r[start:end]
        # Write new FASTA file #
        self.filtered.write(only_16s_portion(self.source))
        # Return #
        return self.filtered</code></pre>
</details>
<h3>Ancestors</h3>
<ul class="hlist">
<li><a title="pacmill.filtering.barrnap.Barrnap" href="#pacmill.filtering.barrnap.Barrnap">Barrnap</a></li>
</ul>
<h3>Methods</h3>
<dl>
<dt id="pacmill.filtering.barrnap.BarrnapExtract.extract"><code class="name flex">
<span>def <span class="ident">extract</span></span>(<span>self, verbose=True)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def extract(self, verbose=True):
    # Message #
    if verbose:
        msg = &#34;Extracting the 16S rRNA portion of sequences from &#39;%s&#39;&#34;
        print(msg % self.dest)
    # Get reads that had a rRNA hit #
    reader = tag.GFF3Reader(infilename=self.dest)
    reader = tag.select.features(reader, type=&#39;rRNA&#39;)
    # Make a lookup table for those with 16S hits #
    lookup = {rec.seqid: (rec.start, rec.end)
              for rec in reader if &#39;16S_rRNA&#39; in rec.attributes}
    # Function to yield only the good part of each read #
    def only_16s_portion(reads):
        for r in reads:
            start_and_end = lookup.get(r.id)
            if start_and_end is None: continue
            start, end = start_and_end
            yield r[start:end]
    # Write new FASTA file #
    self.filtered.write(only_16s_portion(self.source))
    # Return #
    return self.filtered</code></pre>
</details>
</dd>
</dl>
<h3>Inherited members</h3>
<ul class="hlist">
<li><code><b><a title="pacmill.filtering.barrnap.Barrnap" href="#pacmill.filtering.barrnap.Barrnap">Barrnap</a></b></code>:
<ul class="hlist">
<li><code><a title="pacmill.filtering.barrnap.Barrnap.check_installed" href="#pacmill.filtering.barrnap.Barrnap.check_installed">check_installed</a></code></li>
<li><code><a title="pacmill.filtering.barrnap.Barrnap.install" href="#pacmill.filtering.barrnap.Barrnap.install">install</a></code></li>
<li><code><a title="pacmill.filtering.barrnap.Barrnap.results" href="#pacmill.filtering.barrnap.Barrnap.results">results</a></code></li>
</ul>
</li>
</ul>
</dd>
<dt id="pacmill.filtering.barrnap.BarrnapFilter"><code class="flex name class">
<span>class <span class="ident">BarrnapFilter</span></span>
<span>(</span><span>source, dest=None, filtered=None)</span>
</code></dt>
<dd>
<div class="desc"><p>With this subclass, we will filter reads after running barrnap.</p>
<p>Using the original reads file and the GFF output of barrnap,
we will create a new FASTQ file containing only the original reads
that had a hit for both rRNA genes on the same read.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class BarrnapFilter(Barrnap):
    &#34;&#34;&#34;
    With this subclass, we will filter reads after running barrnap.

    Using the original reads file and the GFF output of barrnap,
    we will create a new FASTQ file containing only the original reads
    that had a hit for both rRNA genes on the same read.
    &#34;&#34;&#34;

    def __call__(self, cpus=None, verbose=True):
        # Call the parent class method #
        self.run(cpus, verbose)
        # Filter #
        self.filter()

    #--------------------------- Presence of gene ----------------------------#
    def parse_hit_ids(self, attr_text):
        &#34;&#34;&#34;
        Using the GFF output of barrnap and the `tag` parser, we retrieve
        all IDs that contained a given text in their attributes.
        &#34;&#34;&#34;
        reader = tag.GFF3Reader(infilename=self.dest)
        reader = tag.select.features(reader, type=&#39;rRNA&#39;)
        return [rec.seqid for rec in reader if attr_text in rec.attributes]

    @property_cached
    def ids_16s(self):
        &#34;&#34;&#34;Return read IDs for reads that contained a 16S gene.&#34;&#34;&#34;
        return frozenset(self.parse_hit_ids(&#39;16S_rRNA&#39;))

    @property_cached
    def ids_23s(self):
        &#34;&#34;&#34;Return read IDs for reads that contained a 23S gene.&#34;&#34;&#34;
        return frozenset(self.parse_hit_ids(&#39;23S_rRNA&#39;))

    def filter(self, verbose=True):
        # Message #
        if verbose:
            msg = &#34;Extracting sequences with both rRNA genes from &#39;%s&#39;&#34;
            print(msg % self.dest)
        # Get those that had both genes #
        ids_both = self.ids_16s &amp; self.ids_23s
        # Extract those IDs #
        FASTQ(self.source).extract_sequences(ids_both, self.filtered)
        # Return #
        return self.filtered</code></pre>
</details>
<h3>Ancestors</h3>
<ul class="hlist">
<li><a title="pacmill.filtering.barrnap.Barrnap" href="#pacmill.filtering.barrnap.Barrnap">Barrnap</a></li>
</ul>
<h3>Instance variables</h3>
<dl>
<dt id="pacmill.filtering.barrnap.BarrnapFilter.ids_16s"><code class="name">var <span class="ident">ids_16s</span></code></dt>
<dd>
<div class="desc"><p>Return read IDs for reads that contained a 16S gene.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def __get__(self, instance, owner):
    &#34;&#34;&#34;
    If you see the current source code in a seemingly unrelated part of
    an auto-generated documentation, it means the program making the
    documentation was unable to correctly traverse a decorated property.
    &#34;&#34;&#34;
    # For debugging purposes #
    if False: print(&#34;-&gt; property cached `%s`&#34; % instance)
    # If called from a class #
    if instance is None: return self
    # Does a cache exist for this instance? #
    self.check_cache(instance)
    # Is the answer in the cache? #
    if self.name in instance.__cache__: return instance.__cache__[self.name]
    # If not we will compute it #
    if inspect.isgeneratorfunction(self.func): result = tuple(self.func(instance))
    else:                                      result = self.func(instance)
    # Let&#39;s store the answer for later #
    instance.__cache__[self.name] = result
    # Return #
    return result</code></pre>
</details>
</dd>
<dt id="pacmill.filtering.barrnap.BarrnapFilter.ids_23s"><code class="name">var <span class="ident">ids_23s</span></code></dt>
<dd>
<div class="desc"><p>Return read IDs for reads that contained a 23S gene.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def __get__(self, instance, owner):
    &#34;&#34;&#34;
    If you see the current source code in a seemingly unrelated part of
    an auto-generated documentation, it means the program making the
    documentation was unable to correctly traverse a decorated property.
    &#34;&#34;&#34;
    # For debugging purposes #
    if False: print(&#34;-&gt; property cached `%s`&#34; % instance)
    # If called from a class #
    if instance is None: return self
    # Does a cache exist for this instance? #
    self.check_cache(instance)
    # Is the answer in the cache? #
    if self.name in instance.__cache__: return instance.__cache__[self.name]
    # If not we will compute it #
    if inspect.isgeneratorfunction(self.func): result = tuple(self.func(instance))
    else:                                      result = self.func(instance)
    # Let&#39;s store the answer for later #
    instance.__cache__[self.name] = result
    # Return #
    return result</code></pre>
</details>
</dd>
</dl>
<h3>Methods</h3>
<dl>
<dt id="pacmill.filtering.barrnap.BarrnapFilter.filter"><code class="name flex">
<span>def <span class="ident">filter</span></span>(<span>self, verbose=True)</span>
</code></dt>
<dd>
<div class="desc"></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def filter(self, verbose=True):
    # Message #
    if verbose:
        msg = &#34;Extracting sequences with both rRNA genes from &#39;%s&#39;&#34;
        print(msg % self.dest)
    # Get those that had both genes #
    ids_both = self.ids_16s &amp; self.ids_23s
    # Extract those IDs #
    FASTQ(self.source).extract_sequences(ids_both, self.filtered)
    # Return #
    return self.filtered</code></pre>
</details>
</dd>
<dt id="pacmill.filtering.barrnap.BarrnapFilter.parse_hit_ids"><code class="name flex">
<span>def <span class="ident">parse_hit_ids</span></span>(<span>self, attr_text)</span>
</code></dt>
<dd>
<div class="desc"><p>Using the GFF output of barrnap and the <code>tag</code> parser, we retrieve
all IDs that contained a given text in their attributes.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def parse_hit_ids(self, attr_text):
    &#34;&#34;&#34;
    Using the GFF output of barrnap and the `tag` parser, we retrieve
    all IDs that contained a given text in their attributes.
    &#34;&#34;&#34;
    reader = tag.GFF3Reader(infilename=self.dest)
    reader = tag.select.features(reader, type=&#39;rRNA&#39;)
    return [rec.seqid for rec in reader if attr_text in rec.attributes]</code></pre>
</details>
</dd>
</dl>
<h3>Inherited members</h3>
<ul class="hlist">
<li><code><b><a title="pacmill.filtering.barrnap.Barrnap" href="#pacmill.filtering.barrnap.Barrnap">Barrnap</a></b></code>:
<ul class="hlist">
<li><code><a title="pacmill.filtering.barrnap.Barrnap.check_installed" href="#pacmill.filtering.barrnap.Barrnap.check_installed">check_installed</a></code></li>
<li><code><a title="pacmill.filtering.barrnap.Barrnap.install" href="#pacmill.filtering.barrnap.Barrnap.install">install</a></code></li>
<li><code><a title="pacmill.filtering.barrnap.Barrnap.results" href="#pacmill.filtering.barrnap.Barrnap.results">results</a></code></li>
</ul>
</li>
</ul>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="pacmill.filtering" href="index.html">pacmill.filtering</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="pacmill.filtering.barrnap.Barrnap" href="#pacmill.filtering.barrnap.Barrnap">Barrnap</a></code></h4>
<ul class="two-column">
<li><code><a title="pacmill.filtering.barrnap.Barrnap.apt_packages" href="#pacmill.filtering.barrnap.Barrnap.apt_packages">apt_packages</a></code></li>
<li><code><a title="pacmill.filtering.barrnap.Barrnap.check_installed" href="#pacmill.filtering.barrnap.Barrnap.check_installed">check_installed</a></code></li>
<li><code><a title="pacmill.filtering.barrnap.Barrnap.install" href="#pacmill.filtering.barrnap.Barrnap.install">install</a></code></li>
<li><code><a title="pacmill.filtering.barrnap.Barrnap.reject_threshold" href="#pacmill.filtering.barrnap.Barrnap.reject_threshold">reject_threshold</a></code></li>
<li><code><a title="pacmill.filtering.barrnap.Barrnap.results" href="#pacmill.filtering.barrnap.Barrnap.results">results</a></code></li>
<li><code><a title="pacmill.filtering.barrnap.Barrnap.run" href="#pacmill.filtering.barrnap.Barrnap.run">run</a></code></li>
<li><code><a title="pacmill.filtering.barrnap.Barrnap.zip_url" href="#pacmill.filtering.barrnap.Barrnap.zip_url">zip_url</a></code></li>
</ul>
</li>
<li>
<h4><code><a title="pacmill.filtering.barrnap.BarrnapExtract" href="#pacmill.filtering.barrnap.BarrnapExtract">BarrnapExtract</a></code></h4>
<ul class="">
<li><code><a title="pacmill.filtering.barrnap.BarrnapExtract.extract" href="#pacmill.filtering.barrnap.BarrnapExtract.extract">extract</a></code></li>
</ul>
</li>
<li>
<h4><code><a title="pacmill.filtering.barrnap.BarrnapFilter" href="#pacmill.filtering.barrnap.BarrnapFilter">BarrnapFilter</a></code></h4>
<ul class="">
<li><code><a title="pacmill.filtering.barrnap.BarrnapFilter.filter" href="#pacmill.filtering.barrnap.BarrnapFilter.filter">filter</a></code></li>
<li><code><a title="pacmill.filtering.barrnap.BarrnapFilter.ids_16s" href="#pacmill.filtering.barrnap.BarrnapFilter.ids_16s">ids_16s</a></code></li>
<li><code><a title="pacmill.filtering.barrnap.BarrnapFilter.ids_23s" href="#pacmill.filtering.barrnap.BarrnapFilter.ids_23s">ids_23s</a></code></li>
<li><code><a title="pacmill.filtering.barrnap.BarrnapFilter.parse_hit_ids" href="#pacmill.filtering.barrnap.BarrnapFilter.parse_hit_ids">parse_hit_ids</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.8.4</a>.</p>
</footer>
</body>
</html>