<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.9.2" />
<title>pacmill.centering.vsearch API documentation</title>
<meta name="description" content="Written by Lucas Sinclair.
MIT Licensed.
Contact at www.sinclair.bio" />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>pacmill.centering.vsearch</code></h1>
</header>
<section id="section-intro">
<p>Written by Lucas Sinclair.
MIT Licensed.
Contact at www.sinclair.bio</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">#!/usr/bin/env python3
# -*- coding: utf-8 -*-

&#34;&#34;&#34;
Written by Lucas Sinclair.
MIT Licensed.
Contact at www.sinclair.bio
&#34;&#34;&#34;

# Built-in modules #
import multiprocessing, re, shutil

# First party modules #
from fasta import FASTA, FASTQ
from plumbing.check_cmd_found import check_cmd
from plumbing.cache           import property_cached
from autopaths.file_path      import FilePath

# Third party modules #
import sh, pandas

###############################################################################
class ClusterVsearch:
    &#34;&#34;&#34;
    Takes care of clustering reads into OTUs by calling
    `vsearch --cluster_size`.

    From the man pages:

        Output cluster consensus sequences to filename. For each cluster,
        a multiple alignment is computed, and a consensus sequence is
        constructed by taking the majority symbol (nucleotide or gap) from
        each column of the alignment.

    Caveat: The Vsearch algorithm will name OTU differently in the FASTA
            file it outputs as compared to the TSV file it outputs

    In the `self.table` which is a TSV they are named:

        sample_1:1034      sample_1:1041       etc.

    In the `self.otus` which is a FASTA they are named:

       centroid=sample_1:1034;seqs=128      etc.

    After mothur classifies the OTU, they are named:

       centroid=sample_1_1034;seqs=128      etc.

    Which is almost the same except colons are underscores now.
    &#34;&#34;&#34;

    def __repr__(self):
        msg = &#39;&lt;%s object on &#34;%s&#34;&gt;&#39;
        return msg % (self.__class__.__name__, self.source.path)

    def __init__(self, source, threshold, min_size, otus=None, table=None):
        # Source is a FASTA to run the algorithm on #
        self.source = FASTQ(source)
        # Threshold is a float such as 0.97 #
        self.threshold = threshold
        # Minimum size is an integer such as 2 #
        self.min_size = min_size
        # self.otus is a FASTA file that contains the centroid sequences #
        if otus is None:
            otus = self.source.prefix_path + &#39;.otus.fasta&#39;
        self.otus = FASTA(otus)
        # self.table is a TSV file that contains the per sample counts #
        if table is None:
            table = self.source.prefix_path + &#39;.otus.tsv&#39;
        self.table = FilePath(table)

    #------------------------------ Running ----------------------------------#
    def __call__(self, cpus=None, verbose=True):
        # Call vsearch #
        self.cluster(cpus, verbose)
        # Filter OTUs that are too small #
        if self.min_size &gt; 1: self.dereplicate(verbose)
        # Return #
        return self.otus

    def cluster(self, cpus=None, verbose=True):
        &#34;&#34;&#34;
        Will call the vsearch algorithm. Documentation for command line
        options are found at:
        https://manpages.debian.org/stretch/vsearch/vsearch.1.en.html
        &#34;&#34;&#34;
        # Message #
        if verbose:
            msg = &#34;Running OTU creation on &#39;%s&#39;&#34;
            print(msg % self.source)
        # Check it is installed #
        check_cmd(&#39;vsearch&#39;, True)
        # Number of cores #
        if cpus is None: cpus = min(multiprocessing.cpu_count(), 32)
        # Pick the command parameters #
        command = (&#34;--cluster_size&#34;, self.source,
                   &#34;--consout&#34;,      self.otus,
                   &#34;--id&#34;,           self.threshold,
                   &#34;--otutabout&#34;,    self.table,
                   &#34;--threads&#34;,      cpus)
        # Run the command on the input FASTA file #
        sh.vsearch(command)
        # Return #
        return self.otus

    def dereplicate(self, verbose=True):
        &#34;&#34;&#34;
        We will read the TSV table at `self.table` and remove OTUs that are
        below the self.min_size threshold. Following which, we will read the
        FASTA file at `self.otus` and remove the same sequences that are no
        longer required.
        &#34;&#34;&#34;
        # Message #
        if verbose:
            msg = &#34;Removing low abundance OTUs on &#39;%s&#39;&#34;
            print(msg % self.otus)
        # Load table #
        df = pandas.read_csv(self.table, sep=&#39;\t&#39;, index_col=0)
        # Sum and sort #
        totals = df.sum(axis=1).sort_values(ascending=False)
        # Get the sequences IDs to drop and those to keep #
        self.keep_ids = list(totals[totals &gt;= self.min_size].index)
        self.drop_ids = list(totals[totals &lt;  self.min_size].index)
        # Filter the dataframe #
        df = df.loc[self.keep_ids]
        # Backup the old table and replace it #
        shutil.move(self.table, self.table + &#39;.unfiltered&#39;)
        df.to_csv(self.table.path, sep=&#39;\t&#39;)
        # Function for filtering reads #
        def keep_reads_if(title):
            pattern  = r&#39;\Acentroid=(.+);seqs=[0-9]+\Z&#39;
            otu_name = re.findall(pattern, title)[0]
            if otu_name in self.keep_ids: return True
            return False
        # Filter the FASTA file #
        self.otus.copy(self.otus + &#39;.unfiltered&#39;)
        self.otus.extract_sequences(keep_reads_if, in_place=True)
        # Return #
        return self.otus

    #------------------------------- Results ---------------------------------#
    def __bool__(self):
        &#34;&#34;&#34;
        Return True if the OTU creation software was run already and the
        results are stored on the filesystem. Return False if it was not yet
        run.
        &#34;&#34;&#34;
        return self.otus.exists

    @property_cached
    def results(self):
        # Check it was run #
        if not self:
            msg = &#34;You can&#39;t access results from OTU creation &#34; \
                  &#34;before running the tool.&#34;
            raise Exception(msg)
        # Return the results #
        return ClusteringResults(self.otus)

###############################################################################
class ClusteringResults(FASTA):
    &#34;&#34;&#34;A file with the results.&#34;&#34;&#34;
    pass</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="pacmill.centering.vsearch.ClusterVsearch"><code class="flex name class">
<span>class <span class="ident">ClusterVsearch</span></span>
<span>(</span><span>source, threshold, min_size, otus=None, table=None)</span>
</code></dt>
<dd>
<div class="desc"><p>Takes care of clustering reads into OTUs by calling
<code>vsearch --cluster_size</code>.</p>
<p>From the man pages:</p>
<pre><code>Output cluster consensus sequences to filename. For each cluster,
a multiple alignment is computed, and a consensus sequence is
constructed by taking the majority symbol (nucleotide or gap) from
each column of the alignment.
</code></pre>
<p>Caveat: The Vsearch algorithm will name OTU differently in the FASTA
file it outputs as compared to the TSV file it outputs</p>
<p>In the <code>self.table</code> which is a TSV they are named:</p>
<pre><code>sample_1:1034      sample_1:1041       etc.
</code></pre>
<p>In the <code>self.otus</code> which is a FASTA they are named:</p>
<p>centroid=sample_1:1034;seqs=128
etc.</p>
<p>After mothur classifies the OTU, they are named:</p>
<p>centroid=sample_1_1034;seqs=128
etc.</p>
<p>Which is almost the same except colons are underscores now.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class ClusterVsearch:
    &#34;&#34;&#34;
    Takes care of clustering reads into OTUs by calling
    `vsearch --cluster_size`.

    From the man pages:

        Output cluster consensus sequences to filename. For each cluster,
        a multiple alignment is computed, and a consensus sequence is
        constructed by taking the majority symbol (nucleotide or gap) from
        each column of the alignment.

    Caveat: The Vsearch algorithm will name OTU differently in the FASTA
            file it outputs as compared to the TSV file it outputs

    In the `self.table` which is a TSV they are named:

        sample_1:1034      sample_1:1041       etc.

    In the `self.otus` which is a FASTA they are named:

       centroid=sample_1:1034;seqs=128      etc.

    After mothur classifies the OTU, they are named:

       centroid=sample_1_1034;seqs=128      etc.

    Which is almost the same except colons are underscores now.
    &#34;&#34;&#34;

    def __repr__(self):
        msg = &#39;&lt;%s object on &#34;%s&#34;&gt;&#39;
        return msg % (self.__class__.__name__, self.source.path)

    def __init__(self, source, threshold, min_size, otus=None, table=None):
        # Source is a FASTA to run the algorithm on #
        self.source = FASTQ(source)
        # Threshold is a float such as 0.97 #
        self.threshold = threshold
        # Minimum size is an integer such as 2 #
        self.min_size = min_size
        # self.otus is a FASTA file that contains the centroid sequences #
        if otus is None:
            otus = self.source.prefix_path + &#39;.otus.fasta&#39;
        self.otus = FASTA(otus)
        # self.table is a TSV file that contains the per sample counts #
        if table is None:
            table = self.source.prefix_path + &#39;.otus.tsv&#39;
        self.table = FilePath(table)

    #------------------------------ Running ----------------------------------#
    def __call__(self, cpus=None, verbose=True):
        # Call vsearch #
        self.cluster(cpus, verbose)
        # Filter OTUs that are too small #
        if self.min_size &gt; 1: self.dereplicate(verbose)
        # Return #
        return self.otus

    def cluster(self, cpus=None, verbose=True):
        &#34;&#34;&#34;
        Will call the vsearch algorithm. Documentation for command line
        options are found at:
        https://manpages.debian.org/stretch/vsearch/vsearch.1.en.html
        &#34;&#34;&#34;
        # Message #
        if verbose:
            msg = &#34;Running OTU creation on &#39;%s&#39;&#34;
            print(msg % self.source)
        # Check it is installed #
        check_cmd(&#39;vsearch&#39;, True)
        # Number of cores #
        if cpus is None: cpus = min(multiprocessing.cpu_count(), 32)
        # Pick the command parameters #
        command = (&#34;--cluster_size&#34;, self.source,
                   &#34;--consout&#34;,      self.otus,
                   &#34;--id&#34;,           self.threshold,
                   &#34;--otutabout&#34;,    self.table,
                   &#34;--threads&#34;,      cpus)
        # Run the command on the input FASTA file #
        sh.vsearch(command)
        # Return #
        return self.otus

    def dereplicate(self, verbose=True):
        &#34;&#34;&#34;
        We will read the TSV table at `self.table` and remove OTUs that are
        below the self.min_size threshold. Following which, we will read the
        FASTA file at `self.otus` and remove the same sequences that are no
        longer required.
        &#34;&#34;&#34;
        # Message #
        if verbose:
            msg = &#34;Removing low abundance OTUs on &#39;%s&#39;&#34;
            print(msg % self.otus)
        # Load table #
        df = pandas.read_csv(self.table, sep=&#39;\t&#39;, index_col=0)
        # Sum and sort #
        totals = df.sum(axis=1).sort_values(ascending=False)
        # Get the sequences IDs to drop and those to keep #
        self.keep_ids = list(totals[totals &gt;= self.min_size].index)
        self.drop_ids = list(totals[totals &lt;  self.min_size].index)
        # Filter the dataframe #
        df = df.loc[self.keep_ids]
        # Backup the old table and replace it #
        shutil.move(self.table, self.table + &#39;.unfiltered&#39;)
        df.to_csv(self.table.path, sep=&#39;\t&#39;)
        # Function for filtering reads #
        def keep_reads_if(title):
            pattern  = r&#39;\Acentroid=(.+);seqs=[0-9]+\Z&#39;
            otu_name = re.findall(pattern, title)[0]
            if otu_name in self.keep_ids: return True
            return False
        # Filter the FASTA file #
        self.otus.copy(self.otus + &#39;.unfiltered&#39;)
        self.otus.extract_sequences(keep_reads_if, in_place=True)
        # Return #
        return self.otus

    #------------------------------- Results ---------------------------------#
    def __bool__(self):
        &#34;&#34;&#34;
        Return True if the OTU creation software was run already and the
        results are stored on the filesystem. Return False if it was not yet
        run.
        &#34;&#34;&#34;
        return self.otus.exists

    @property_cached
    def results(self):
        # Check it was run #
        if not self:
            msg = &#34;You can&#39;t access results from OTU creation &#34; \
                  &#34;before running the tool.&#34;
            raise Exception(msg)
        # Return the results #
        return ClusteringResults(self.otus)</code></pre>
</details>
<h3>Instance variables</h3>
<dl>
<dt id="pacmill.centering.vsearch.ClusterVsearch.results"><code class="name">var <span class="ident">results</span></code></dt>
<dd>
<div class="desc"><p>If you see the current source code in a seemingly unrelated part of
an auto-generated documentation, it means the program making the
documentation was unable to correctly traverse a decorated property.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def __get__(self, instance, owner):
    &#34;&#34;&#34;
    If you see the current source code in a seemingly unrelated part of
    an auto-generated documentation, it means the program making the
    documentation was unable to correctly traverse a decorated property.
    &#34;&#34;&#34;
    # For debugging purposes #
    if False: print(&#34;-&gt; property cached `%s`&#34; % instance)
    # If called from a class #
    if instance is None: return self
    # Does a cache exist for this instance? #
    self.check_cache(instance)
    # Is the answer in the cache? #
    if self.name in instance.__cache__: return instance.__cache__[self.name]
    # If not we will compute it #
    if inspect.isgeneratorfunction(self.func): result = tuple(self.func(instance))
    else:                                      result = self.func(instance)
    # Let&#39;s store the answer for later #
    instance.__cache__[self.name] = result
    # Return #
    return result</code></pre>
</details>
</dd>
</dl>
<h3>Methods</h3>
<dl>
<dt id="pacmill.centering.vsearch.ClusterVsearch.cluster"><code class="name flex">
<span>def <span class="ident">cluster</span></span>(<span>self, cpus=None, verbose=True)</span>
</code></dt>
<dd>
<div class="desc"><p>Will call the vsearch algorithm. Documentation for command line
options are found at:
<a href="https://manpages.debian.org/stretch/vsearch/vsearch.1.en.html">https://manpages.debian.org/stretch/vsearch/vsearch.1.en.html</a></p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def cluster(self, cpus=None, verbose=True):
    &#34;&#34;&#34;
    Will call the vsearch algorithm. Documentation for command line
    options are found at:
    https://manpages.debian.org/stretch/vsearch/vsearch.1.en.html
    &#34;&#34;&#34;
    # Message #
    if verbose:
        msg = &#34;Running OTU creation on &#39;%s&#39;&#34;
        print(msg % self.source)
    # Check it is installed #
    check_cmd(&#39;vsearch&#39;, True)
    # Number of cores #
    if cpus is None: cpus = min(multiprocessing.cpu_count(), 32)
    # Pick the command parameters #
    command = (&#34;--cluster_size&#34;, self.source,
               &#34;--consout&#34;,      self.otus,
               &#34;--id&#34;,           self.threshold,
               &#34;--otutabout&#34;,    self.table,
               &#34;--threads&#34;,      cpus)
    # Run the command on the input FASTA file #
    sh.vsearch(command)
    # Return #
    return self.otus</code></pre>
</details>
</dd>
<dt id="pacmill.centering.vsearch.ClusterVsearch.dereplicate"><code class="name flex">
<span>def <span class="ident">dereplicate</span></span>(<span>self, verbose=True)</span>
</code></dt>
<dd>
<div class="desc"><p>We will read the TSV table at <code>self.table</code> and remove OTUs that are
below the self.min_size threshold. Following which, we will read the
FASTA file at <code>self.otus</code> and remove the same sequences that are no
longer required.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def dereplicate(self, verbose=True):
    &#34;&#34;&#34;
    We will read the TSV table at `self.table` and remove OTUs that are
    below the self.min_size threshold. Following which, we will read the
    FASTA file at `self.otus` and remove the same sequences that are no
    longer required.
    &#34;&#34;&#34;
    # Message #
    if verbose:
        msg = &#34;Removing low abundance OTUs on &#39;%s&#39;&#34;
        print(msg % self.otus)
    # Load table #
    df = pandas.read_csv(self.table, sep=&#39;\t&#39;, index_col=0)
    # Sum and sort #
    totals = df.sum(axis=1).sort_values(ascending=False)
    # Get the sequences IDs to drop and those to keep #
    self.keep_ids = list(totals[totals &gt;= self.min_size].index)
    self.drop_ids = list(totals[totals &lt;  self.min_size].index)
    # Filter the dataframe #
    df = df.loc[self.keep_ids]
    # Backup the old table and replace it #
    shutil.move(self.table, self.table + &#39;.unfiltered&#39;)
    df.to_csv(self.table.path, sep=&#39;\t&#39;)
    # Function for filtering reads #
    def keep_reads_if(title):
        pattern  = r&#39;\Acentroid=(.+);seqs=[0-9]+\Z&#39;
        otu_name = re.findall(pattern, title)[0]
        if otu_name in self.keep_ids: return True
        return False
    # Filter the FASTA file #
    self.otus.copy(self.otus + &#39;.unfiltered&#39;)
    self.otus.extract_sequences(keep_reads_if, in_place=True)
    # Return #
    return self.otus</code></pre>
</details>
</dd>
</dl>
</dd>
<dt id="pacmill.centering.vsearch.ClusteringResults"><code class="flex name class">
<span>class <span class="ident">ClusteringResults</span></span>
<span>(</span><span>path)</span>
</code></dt>
<dd>
<div class="desc"><p>A file with the results.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class ClusteringResults(FASTA):
    &#34;&#34;&#34;A file with the results.&#34;&#34;&#34;
    pass</code></pre>
</details>
<h3>Ancestors</h3>
<ul class="hlist">
<li>fasta.core.FASTA</li>
<li>autopaths.file_path.FilePath</li>
<li>autopaths.base_path.BasePath</li>
<li>builtins.str</li>
</ul>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="pacmill.centering" href="index.html">pacmill.centering</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="pacmill.centering.vsearch.ClusterVsearch" href="#pacmill.centering.vsearch.ClusterVsearch">ClusterVsearch</a></code></h4>
<ul class="">
<li><code><a title="pacmill.centering.vsearch.ClusterVsearch.cluster" href="#pacmill.centering.vsearch.ClusterVsearch.cluster">cluster</a></code></li>
<li><code><a title="pacmill.centering.vsearch.ClusterVsearch.dereplicate" href="#pacmill.centering.vsearch.ClusterVsearch.dereplicate">dereplicate</a></code></li>
<li><code><a title="pacmill.centering.vsearch.ClusterVsearch.results" href="#pacmill.centering.vsearch.ClusterVsearch.results">results</a></code></li>
</ul>
</li>
<li>
<h4><code><a title="pacmill.centering.vsearch.ClusteringResults" href="#pacmill.centering.vsearch.ClusteringResults">ClusteringResults</a></code></h4>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc"><cite>pdoc</cite> 0.9.2</a>.</p>
</footer>
</body>
</html>